{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "42afb47b-86c5-465d-8127-e5ab69643813",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import gzip\n",
    "import numpy as np\n",
    "from itertools import product\n",
    "from joblib import Parallel, delayed\n",
    "from tqdm.auto import tqdm\n",
    "from collections import defaultdict, Counter, OrderedDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b36aadef-6d9a-4789-914f-ed746309cd15",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to read FASTA file and yield name and sequence\n",
    "def fasta_iter(fasta_file):\n",
    "    with gzip.open(fasta_file, 'rt') as file:\n",
    "        name = None\n",
    "        seqs = []\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\">\"):\n",
    "                if name:\n",
    "                    yield name, ''.join(seqs)\n",
    "                name = line[1:]  # Skip '>'\n",
    "                seqs = []\n",
    "            else:\n",
    "                seqs.append(line)\n",
    "        if name:\n",
    "            yield name, ''.join(seqs)  # Yield the last entry\n",
    "\n",
    "# Function to read FASTQ file and yield name and sequence\n",
    "def fastq_iter(fastq_file):\n",
    "    with gzip.open(fastq_file, 'rt') as file:\n",
    "        while True:\n",
    "            name = file.readline().strip()[1:]  # Skip '@' in FASTQ\n",
    "            seq = file.readline().strip()\n",
    "            file.readline()  # Skip the '+' line\n",
    "            file.readline()  # Skip the quality line\n",
    "            if not name or not seq:\n",
    "                break\n",
    "            yield name, seq\n",
    "\n",
    "# Function to generate feature mapping for k-mers with sequences as keys\n",
    "def generate_feature_mapping(kmer_len):\n",
    "    BASE_COMPLEMENT = {\"A\": \"T\", \"T\": \"A\", \"G\": \"C\", \"C\": \"G\"}\n",
    "    kmer_hash = {}\n",
    "    kmer_names = []\n",
    "    counter = 0\n",
    "    for kmer in product(\"ATGC\", repeat=kmer_len):\n",
    "        kmer = ''.join(kmer)\n",
    "        rev_compl = ''.join([BASE_COMPLEMENT[x] for x in reversed(kmer)])  # Generate reverse complement\n",
    "        if kmer not in kmer_hash and rev_compl not in kmer_hash:\n",
    "            # Use canonical k-mer (lexicographically smaller of kmer and rev_compl)\n",
    "            canonical_kmer = min(kmer, rev_compl)\n",
    "            kmer_hash[canonical_kmer] = counter\n",
    "            kmer_names.append(canonical_kmer)  # Store the k-mer for naming features\n",
    "            counter += 1\n",
    "    return kmer_hash, kmer_names, counter\n",
    "\n",
    "# Fungsi untuk menghitung frekuensi global k-mer\n",
    "def compute_global_kmer_abundance(seq_iter, kmer_len):\n",
    "    global_kmer_counts = Counter()\n",
    "    for h, seq in seq_iter:\n",
    "        norm_seq = str(seq).upper()\n",
    "        for i in range(len(norm_seq) - kmer_len + 1):\n",
    "            kmer = norm_seq[i:i+kmer_len]\n",
    "            if set(kmer) <= {'A', 'T', 'G', 'C'}:\n",
    "                global_kmer_counts[kmer] += 1\n",
    "    return global_kmer_counts\n",
    "    \n",
    "# Function to detect file type based on file extension\n",
    "def detect_file_type(file_path):\n",
    "    ext = os.path.splitext(file_path)[-1].lower()  # Get file extension and convert to lowercase\n",
    "    # Handle gzipped files\n",
    "    if ext == '.gz':\n",
    "        ext = os.path.splitext(file_path[:-3])[-1].lower()  # Check the extension before \".gz\"\n",
    "        \n",
    "    if ext in ['.fasta', '.fa', '.fna']:\n",
    "        return 'fasta'\n",
    "    elif ext in ['.fastq', '.fq']:\n",
    "        return 'fastq'\n",
    "    else:\n",
    "        raise ValueError(\"Unsupported file extension. Please provide a valid FASTA or FASTQ file.\")\n",
    "        \n",
    "def generate_kmer_features(file_path, length_threshold, kmer_len, split=False, split_threshold=0):\n",
    "    # Deteksi tipe file dan inisialisasi iterator\n",
    "    file_type = detect_file_type(file_path)\n",
    "    if file_type == 'fasta':\n",
    "        seq_iter = fasta_iter(file_path)\n",
    "    elif file_type == 'fastq':\n",
    "        seq_iter = fastq_iter(file_path)\n",
    "    else:\n",
    "        raise ValueError(\"Unsupported file type. Use 'fasta' or 'fastq'.\")\n",
    "\n",
    "    # Hitung abundance global k-mer\n",
    "    global_kmer_counts = compute_global_kmer_abundance(seq_iter, kmer_len)\n",
    "    \n",
    "    # Reset iterator setelah menghitung abundance global\n",
    "    if file_type == 'fasta':\n",
    "        seq_iter = fasta_iter(file_path)\n",
    "    elif file_type == 'fastq':\n",
    "        seq_iter = fastq_iter(file_path)\n",
    "\n",
    "    def seq_list():\n",
    "        for h, seq in seq_iter:\n",
    "            if not split:\n",
    "                yield h, seq\n",
    "            elif len(seq) >= split_threshold:\n",
    "                half = len(seq) // 2\n",
    "                yield (h + '_1', seq[:half])\n",
    "                yield (h + '_2', seq[half:])\n",
    "\n",
    "    kmer_dict, kmer_names, nr_features = generate_feature_mapping(kmer_len)\n",
    "    composition = OrderedDict()\n",
    "\n",
    "    for h, seq in seq_list():\n",
    "        if len(seq) < length_threshold:\n",
    "            continue\n",
    "        norm_seq = str(seq).upper()\n",
    "        kmers = [kmer_dict[norm_seq[i:i+kmer_len]]\n",
    "                 for i in range(len(norm_seq) - kmer_len + 1)\n",
    "                 if norm_seq[i:i+kmer_len] in kmer_dict]  # Ignore kmers with non-canonical bases\n",
    "        composition[h] = np.bincount(np.array(kmers, dtype=np.int64), minlength=nr_features)\n",
    "\n",
    "    # Create DataFrame with k-mer names as column headers\n",
    "    df = pd.DataFrame.from_dict(composition, orient='index', dtype=float, columns=kmer_names)\n",
    "\n",
    "    df = df.apply(lambda x: x + 1e-5)  # Smoothing\n",
    "    df = df.div(df.sum(axis=1), axis=0)  # Normalize by row (sequence)\n",
    "    \n",
    "    # Tambahkan abundance global ke DataFrame\n",
    "    def compute_seq_abundance(row):\n",
    "        return sum([global_kmer_counts.get(kmer, 0) * row[kmer] for kmer in kmer_names])\n",
    "\n",
    "    df['global_abundance'] = df.apply(compute_seq_abundance, axis=1)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e8f6389-537c-4390-877a-8f960d1474ae",
   "metadata": {},
   "source": [
    "# Contigs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35ad17f4-ea82-4c31-b4ef-fad1fa8a64b8",
   "metadata": {},
   "source": [
    "## Feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "149aa058-c277-4dc2-9fe1-582b4a4aba59",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_file = './contigs/anonymous_gsa.fasta.gz'  # Updated file path to your uploaded file\n",
    "kmer_len = 5  # Original k-mer length\n",
    "length_threshold = 100  # Example sequence length threshold\n",
    "\n",
    "# Generate spaced k-mer features\n",
    "kmer_df = generate_kmer_features(fasta_file, length_threshold, kmer_len)\n",
    "print(kmer_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b030146f-3502-49d6-ad5a-1e5170395687",
   "metadata": {},
   "source": [
    "## Label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ccc77d20-9b15-4645-9fb7-887c1bfa9ebc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Genome4.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genome8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Genome10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Genome6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Genome23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47884</th>\n",
       "      <td>Genome18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47885</th>\n",
       "      <td>Genome10.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47886</th>\n",
       "      <td>Genome8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47887</th>\n",
       "      <td>Genome3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47888</th>\n",
       "      <td>Genome17.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>47889 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        genome_id\n",
       "0       Genome4.1\n",
       "1       Genome8.0\n",
       "2      Genome10.0\n",
       "3       Genome6.0\n",
       "4      Genome23.0\n",
       "...           ...\n",
       "47884  Genome18.0\n",
       "47885  Genome10.1\n",
       "47886   Genome8.0\n",
       "47887   Genome3.0\n",
       "47888  Genome17.0\n",
       "\n",
       "[47889 rows x 1 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "label_path = \"./contigs/gsa_mapping.tsv.gz\"\n",
    "label_df = pd.read_csv(label_path, sep='\\t')\n",
    "label_df = label_df[[\"genome_id\"]]\n",
    "label_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2e12dff7-c5fe-4b9c-8a3e-7240c214b8d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Statistik Deskriptif Data Fitur:\n",
      "               AAAA          AAAT          AAAG          AAAC          AATA  \\\n",
      "count  4.788900e+04  4.788900e+04  4.788900e+04  4.788900e+04  4.788900e+04   \n",
      "mean   1.301096e-02  8.710589e-03  9.223190e-03  7.123879e-03  6.970468e-03   \n",
      "std    1.817247e-02  1.121281e-02  1.026314e-02  8.313860e-03  1.031831e-02   \n",
      "min    1.633983e-08  1.589822e-08  1.680668e-08  1.592353e-08  1.501498e-08   \n",
      "25%    7.999913e-08  7.873931e-08  9.259143e-08  8.333239e-08  7.299198e-08   \n",
      "50%    6.410265e-03  5.649731e-03  7.067139e-03  5.952387e-03  1.470559e-07   \n",
      "75%    1.999983e-02  1.408438e-02  1.449268e-02  1.204815e-02  1.136357e-02   \n",
      "max    1.588766e-01  1.038944e-01  7.058722e-02  9.999748e-02  1.037027e-01   \n",
      "\n",
      "               AATT          AATG          AATC          AAGA          AAGT  \\\n",
      "count  4.788900e+04  4.788900e+04  4.788900e+04  4.788900e+04  4.788900e+04   \n",
      "mean   7.247433e-03  6.220398e-03  6.672922e-03  8.610312e-03  5.611395e-03   \n",
      "std    1.020433e-02  7.837312e-03  7.944416e-03  9.566136e-03  7.252576e-03   \n",
      "min    1.282049e-08  1.472751e-08  1.582275e-08  1.589822e-08  1.016259e-08   \n",
      "25%    7.407333e-08  7.692227e-08  8.064428e-08  9.433841e-08  7.633509e-08   \n",
      "50%    1.724098e-07  4.155128e-03  5.494519e-03  6.896556e-03  3.300348e-03   \n",
      "75%    1.212553e-02  1.052630e-02  1.123594e-02  1.351340e-02  8.781557e-03   \n",
      "max    9.615152e-02  8.536456e-02  8.107837e-02  8.064463e-02  6.249883e-02   \n",
      "\n",
      "       ...          CATG          CAGG          CACG          CTAG  \\\n",
      "count  ...  4.788900e+04  4.788900e+04  4.788900e+04  4.788900e+04   \n",
      "mean   ...  7.193134e-03  9.746998e-03  6.609394e-03  1.493216e-03   \n",
      "std    ...  7.901555e-03  9.934237e-03  8.007340e-03  3.758235e-03   \n",
      "min    ...  1.209188e-08  1.264220e-08  9.115759e-09  8.032120e-09   \n",
      "25%    ...  9.008899e-08  1.282029e-07  8.064428e-08  6.369372e-08   \n",
      "50%    ...  6.329123e-03  7.633585e-03  5.208349e-03  7.873931e-08   \n",
      "75%    ...  1.162787e-02  1.442303e-02  1.092895e-02  1.351327e-07   \n",
      "max    ...  8.029125e-02  1.166642e-01  9.999748e-02  4.761818e-02   \n",
      "\n",
      "               CCAG          CGAG          CCCG          CGCG          CCGG  \\\n",
      "count  4.788900e+04  4.788900e+04  4.788900e+04  4.788900e+04  4.788900e+04   \n",
      "mean   9.950620e-03  8.604771e-03  1.021155e-02  1.226625e-02  1.317319e-02   \n",
      "std    1.041125e-02  1.037622e-02  1.190825e-02  1.568077e-02  1.353117e-02   \n",
      "min    1.303779e-08  5.630626e-09  6.093840e-09  7.513140e-09  6.675561e-09   \n",
      "25%    1.219492e-07  8.620589e-08  9.708610e-08  8.264370e-08  1.470559e-07   \n",
      "50%    7.633585e-03  6.289317e-03  7.092201e-03  6.993010e-03  9.999982e-03   \n",
      "75%    1.470581e-02  1.351340e-02  1.538445e-02  1.904747e-02  1.986744e-02   \n",
      "max    1.111085e-01  1.406222e-01  9.900867e-02  1.645543e-01  1.718715e-01   \n",
      "\n",
      "       global_abundance  \n",
      "count      47889.000000  \n",
      "mean       93076.299758  \n",
      "std        10175.531502  \n",
      "min        62672.960252  \n",
      "25%        85132.540239  \n",
      "50%        92176.019769  \n",
      "75%       100397.153865  \n",
      "max       138759.716245  \n",
      "\n",
      "[8 rows x 137 columns]\n",
      "\n",
      "Distribusi Kelas pada Data Label:\n",
      "genome_id \n",
      "Genome22.0    5764\n",
      "Genome3.0     4911\n",
      "Genome8.0     4667\n",
      "Genome2.0     4496\n",
      "Genome10.1    4344\n",
      "Genome6.0     3546\n",
      "Genome4.0     3188\n",
      "Genome15.0    3087\n",
      "Genome14.0    2698\n",
      "Genome18.0    1813\n",
      "Genome10.0    1788\n",
      "Genome23.0    1591\n",
      "Genome13.0    1338\n",
      "Genome17.0    1229\n",
      "Genome4.1     1002\n",
      "Genome20.0     585\n",
      "Genome7.0      436\n",
      "Genome12.0     316\n",
      "Genome24.0     309\n",
      "Genome9.0      265\n",
      "Genome16.0     247\n",
      "Genome5.0      242\n",
      "Genome11.0      26\n",
      "Genome19.0       1\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Tampilkan statistik deskriptif dari data fitur\n",
    "print(\"Statistik Deskriptif Data Fitur:\")\n",
    "print(kmer_df.describe())\n",
    "\n",
    "# Tampilkan distribusi kelas dari data label\n",
    "print(\"\\nDistribusi Kelas pada Data Label:\")\n",
    "print(label_df.value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "062a483a-03d4-490d-9f85-c500982cf6f1",
   "metadata": {},
   "source": [
    "## Classifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d96efc46-054f-46eb-80e1-d7093195c0be",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "import time\n",
    "from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from sklearn.neural_network import MLPClassifier\n",
    "from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, classification_report"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7517d6b0-cbd9-494b-ab9a-3941c5270bf1",
   "metadata": {},
   "source": [
    "### Pembagian Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2f17964d-9334-4e9f-b9dc-c8fc1a264e3c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Jumlah data latih per kelas:\n",
      "(array(['Genome10.0', 'Genome10.1', 'Genome11.0', 'Genome12.0',\n",
      "       'Genome13.0', 'Genome14.0', 'Genome15.0', 'Genome16.0',\n",
      "       'Genome17.0', 'Genome18.0', 'Genome19.0', 'Genome2.0',\n",
      "       'Genome20.0', 'Genome22.0', 'Genome23.0', 'Genome24.0',\n",
      "       'Genome3.0', 'Genome4.0', 'Genome4.1', 'Genome5.0', 'Genome6.0',\n",
      "       'Genome7.0', 'Genome8.0', 'Genome9.0'], dtype=object), array([1408, 3498,   23,  257, 1061, 2160, 2456,  193,  971, 1457,    1,\n",
      "       3613,  475, 4622, 1265,  250, 3906, 2526,  822,  198, 2845,  331,\n",
      "       3775,  198]))\n",
      "\n",
      "Jumlah data uji per kelas:\n",
      "(array(['Genome10.0', 'Genome10.1', 'Genome11.0', 'Genome12.0',\n",
      "       'Genome13.0', 'Genome14.0', 'Genome15.0', 'Genome16.0',\n",
      "       'Genome17.0', 'Genome18.0', 'Genome2.0', 'Genome20.0',\n",
      "       'Genome22.0', 'Genome23.0', 'Genome24.0', 'Genome3.0', 'Genome4.0',\n",
      "       'Genome4.1', 'Genome5.0', 'Genome6.0', 'Genome7.0', 'Genome8.0',\n",
      "       'Genome9.0'], dtype=object), array([ 380,  846,    3,   59,  277,  538,  631,   54,  258,  356,  883,\n",
      "        110, 1142,  326,   59, 1005,  662,  180,   44,  701,  105,  892,\n",
      "         67]))\n"
     ]
    }
   ],
   "source": [
    "features = kmer_df\n",
    "labels = label_df\n",
    "\n",
    "# Pisahkan data menjadi training dan testing set (80% training, 20% testing)\n",
    "X_train, X_test, y_train, y_test = train_test_split(features, labels, test_size=0.2, random_state=42)\n",
    "\n",
    "print(\"Jumlah data latih per kelas:\")\n",
    "print(np.unique(y_train, return_counts=True))\n",
    "\n",
    "print(\"\\nJumlah data uji per kelas:\")\n",
    "print(np.unique(y_test, return_counts=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e81fefcd-0a21-4c92-9d9b-0716da49763f",
   "metadata": {},
   "source": [
    "### Random Forest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e9b159a5-3c65-4175-b48b-c26bf0cf0712",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/sklearn/base.py:1473: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().\n",
      "  return fit_method(estimator, *args, **kwargs)\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Classification Report untuk Random Forest:\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "  Genome10.0       0.00      0.00      0.00       380\n",
      "  Genome10.1       0.44      0.82      0.57       846\n",
      "  Genome11.0       0.00      0.00      0.00         3\n",
      "  Genome12.0       0.00      0.00      0.00        59\n",
      "  Genome13.0       0.83      0.07      0.13       277\n",
      "  Genome14.0       0.60      0.77      0.68       538\n",
      "  Genome15.0       0.74      0.47      0.57       631\n",
      "  Genome16.0       0.00      0.00      0.00        54\n",
      "  Genome17.0       0.88      0.19      0.32       258\n",
      "  Genome18.0       0.94      0.78      0.85       356\n",
      "   Genome2.0       0.56      0.61      0.58       883\n",
      "  Genome20.0       0.75      0.03      0.05       110\n",
      "  Genome22.0       0.56      0.78      0.65      1142\n",
      "  Genome23.0       0.66      0.19      0.30       326\n",
      "  Genome24.0       0.00      0.00      0.00        59\n",
      "   Genome3.0       0.56      0.73      0.63      1005\n",
      "   Genome4.0       0.66      0.82      0.73       662\n",
      "   Genome4.1       0.00      0.00      0.00       180\n",
      "   Genome5.0       0.00      0.00      0.00        44\n",
      "   Genome6.0       0.65      0.61      0.63       701\n",
      "   Genome7.0       0.00      0.00      0.00       105\n",
      "   Genome8.0       0.71      0.84      0.77       892\n",
      "   Genome9.0       0.73      0.12      0.21        67\n",
      "\n",
      "    accuracy                           0.60      9578\n",
      "   macro avg       0.45      0.34      0.33      9578\n",
      "weighted avg       0.58      0.60      0.55      9578\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/utils/validation.py:1339: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().\n",
      "  y = column_or_1d(y, warn=True)\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):\n",
      "STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.\n",
      "\n",
      "Increase the number of iterations (max_iter) or scale the data as shown in:\n",
      "    https://scikit-learn.org/stable/modules/preprocessing.html\n",
      "Please also refer to the documentation for alternative solver options:\n",
      "    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression\n",
      "  n_iter_i = _check_optimize_result(\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Classification Report untuk Logistic Regression:\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "  Genome10.0       0.00      0.00      0.00       380\n",
      "  Genome10.1       0.00      0.00      0.00       846\n",
      "  Genome11.0       0.00      0.00      0.00         3\n",
      "  Genome12.0       0.00      0.00      0.00        59\n",
      "  Genome13.0       0.00      0.00      0.00       277\n",
      "  Genome14.0       0.00      0.00      0.00       538\n",
      "  Genome15.0       0.00      0.00      0.00       631\n",
      "  Genome16.0       0.00      0.00      0.00        54\n",
      "  Genome17.0       0.00      0.00      0.00       258\n",
      "  Genome18.0       0.00      0.00      0.00       356\n",
      "   Genome2.0       0.00      0.00      0.00       883\n",
      "  Genome20.0       0.00      0.00      0.00       110\n",
      "  Genome22.0       0.12      1.00      0.21      1142\n",
      "  Genome23.0       0.00      0.00      0.00       326\n",
      "  Genome24.0       0.00      0.00      0.00        59\n",
      "   Genome3.0       0.00      0.00      0.00      1005\n",
      "   Genome4.0       0.00      0.00      0.00       662\n",
      "   Genome4.1       0.00      0.00      0.00       180\n",
      "   Genome5.0       0.00      0.00      0.00        44\n",
      "   Genome6.0       0.00      0.00      0.00       701\n",
      "   Genome7.0       0.00      0.00      0.00       105\n",
      "   Genome8.0       0.00      0.00      0.00       892\n",
      "   Genome9.0       0.00      0.00      0.00        67\n",
      "\n",
      "    accuracy                           0.12      9578\n",
      "   macro avg       0.01      0.04      0.01      9578\n",
      "weighted avg       0.01      0.12      0.03      9578\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/utils/validation.py:1339: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().\n",
      "  y = column_or_1d(y, warn=True)\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Classification Report untuk Support Vector Machine:\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "  Genome10.0       0.00      0.00      0.00       380\n",
      "  Genome10.1       0.24      0.66      0.36       846\n",
      "  Genome11.0       0.00      0.00      0.00         3\n",
      "  Genome12.0       0.00      0.00      0.00        59\n",
      "  Genome13.0       0.00      0.00      0.00       277\n",
      "  Genome14.0       0.00      0.00      0.00       538\n",
      "  Genome15.0       0.00      0.00      0.00       631\n",
      "  Genome16.0       0.00      0.00      0.00        54\n",
      "  Genome17.0       0.00      0.00      0.00       258\n",
      "  Genome18.0       0.00      0.00      0.00       356\n",
      "   Genome2.0       0.23      0.11      0.15       883\n",
      "  Genome20.0       0.00      0.00      0.00       110\n",
      "  Genome22.0       0.26      0.61      0.36      1142\n",
      "  Genome23.0       0.00      0.00      0.00       326\n",
      "  Genome24.0       0.00      0.00      0.00        59\n",
      "   Genome3.0       0.26      0.51      0.34      1005\n",
      "   Genome4.0       0.42      0.36      0.39       662\n",
      "   Genome4.1       0.00      0.00      0.00       180\n",
      "   Genome5.0       0.00      0.00      0.00        44\n",
      "   Genome6.0       0.26      0.45      0.33       701\n",
      "   Genome7.0       0.00      0.00      0.00       105\n",
      "   Genome8.0       0.23      0.10      0.14       892\n",
      "   Genome9.0       0.00      0.00      0.00        67\n",
      "\n",
      "    accuracy                           0.26      9578\n",
      "   macro avg       0.08      0.12      0.09      9578\n",
      "weighted avg       0.17      0.26      0.19      9578\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/neighbors/_classification.py:238: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().\n",
      "  return self._fit(X, y)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Classification Report untuk K-Nearest Neighbors:\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "  Genome10.0       0.09      0.14      0.11       380\n",
      "  Genome10.1       0.19      0.33      0.24       846\n",
      "  Genome11.0       0.00      0.00      0.00         3\n",
      "  Genome12.0       0.00      0.00      0.00        59\n",
      "  Genome13.0       0.05      0.04      0.04       277\n",
      "  Genome14.0       0.18      0.18      0.18       538\n",
      "  Genome15.0       0.12      0.15      0.13       631\n",
      "  Genome16.0       0.00      0.00      0.00        54\n",
      "  Genome17.0       0.06      0.05      0.05       258\n",
      "  Genome18.0       0.15      0.13      0.14       356\n",
      "   Genome2.0       0.18      0.18      0.18       883\n",
      "  Genome20.0       0.04      0.01      0.02       110\n",
      "  Genome22.0       0.23      0.24      0.24      1142\n",
      "  Genome23.0       0.13      0.06      0.09       326\n",
      "  Genome24.0       0.00      0.00      0.00        59\n",
      "   Genome3.0       0.25      0.26      0.25      1005\n",
      "   Genome4.0       0.31      0.33      0.32       662\n",
      "   Genome4.1       0.09      0.05      0.07       180\n",
      "   Genome5.0       0.00      0.00      0.00        44\n",
      "   Genome6.0       0.22      0.18      0.20       701\n",
      "   Genome7.0       0.00      0.00      0.00       105\n",
      "   Genome8.0       0.19      0.14      0.16       892\n",
      "   Genome9.0       0.24      0.12      0.16        67\n",
      "\n",
      "    accuracy                           0.19      9578\n",
      "   macro avg       0.12      0.11      0.11      9578\n",
      "weighted avg       0.18      0.19      0.18      9578\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/sklearn/base.py:1473: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().\n",
      "  return fit_method(estimator, *args, **kwargs)\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Classification Report untuk Random Forest:\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "  Genome10.0       0.00      0.00      0.00       380\n",
      "  Genome10.1       0.44      0.82      0.57       846\n",
      "  Genome11.0       0.00      0.00      0.00         3\n",
      "  Genome12.0       0.00      0.00      0.00        59\n",
      "  Genome13.0       0.83      0.07      0.13       277\n",
      "  Genome14.0       0.60      0.77      0.68       538\n",
      "  Genome15.0       0.74      0.47      0.57       631\n",
      "  Genome16.0       0.00      0.00      0.00        54\n",
      "  Genome17.0       0.88      0.19      0.32       258\n",
      "  Genome18.0       0.94      0.78      0.85       356\n",
      "   Genome2.0       0.56      0.61      0.58       883\n",
      "  Genome20.0       0.75      0.03      0.05       110\n",
      "  Genome22.0       0.56      0.78      0.65      1142\n",
      "  Genome23.0       0.66      0.19      0.30       326\n",
      "  Genome24.0       0.00      0.00      0.00        59\n",
      "   Genome3.0       0.56      0.73      0.63      1005\n",
      "   Genome4.0       0.66      0.82      0.73       662\n",
      "   Genome4.1       0.00      0.00      0.00       180\n",
      "   Genome5.0       0.00      0.00      0.00        44\n",
      "   Genome6.0       0.65      0.61      0.63       701\n",
      "   Genome7.0       0.00      0.00      0.00       105\n",
      "   Genome8.0       0.71      0.84      0.77       892\n",
      "   Genome9.0       0.73      0.12      0.21        67\n",
      "\n",
      "    accuracy                           0.60      9578\n",
      "   macro avg       0.45      0.34      0.33      9578\n",
      "weighted avg       0.58      0.60      0.55      9578\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/utils/validation.py:1339: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().\n",
      "  y = column_or_1d(y, warn=True)\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):\n",
      "STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.\n",
      "\n",
      "Increase the number of iterations (max_iter) or scale the data as shown in:\n",
      "    https://scikit-learn.org/stable/modules/preprocessing.html\n",
      "Please also refer to the documentation for alternative solver options:\n",
      "    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression\n",
      "  n_iter_i = _check_optimize_result(\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Classification Report untuk Logistic Regression:\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "  Genome10.0       0.00      0.00      0.00       380\n",
      "  Genome10.1       0.00      0.00      0.00       846\n",
      "  Genome11.0       0.00      0.00      0.00         3\n",
      "  Genome12.0       0.00      0.00      0.00        59\n",
      "  Genome13.0       0.00      0.00      0.00       277\n",
      "  Genome14.0       0.00      0.00      0.00       538\n",
      "  Genome15.0       0.00      0.00      0.00       631\n",
      "  Genome16.0       0.00      0.00      0.00        54\n",
      "  Genome17.0       0.00      0.00      0.00       258\n",
      "  Genome18.0       0.00      0.00      0.00       356\n",
      "   Genome2.0       0.00      0.00      0.00       883\n",
      "  Genome20.0       0.00      0.00      0.00       110\n",
      "  Genome22.0       0.12      1.00      0.21      1142\n",
      "  Genome23.0       0.00      0.00      0.00       326\n",
      "  Genome24.0       0.00      0.00      0.00        59\n",
      "   Genome3.0       0.00      0.00      0.00      1005\n",
      "   Genome4.0       0.00      0.00      0.00       662\n",
      "   Genome4.1       0.00      0.00      0.00       180\n",
      "   Genome5.0       0.00      0.00      0.00        44\n",
      "   Genome6.0       0.00      0.00      0.00       701\n",
      "   Genome7.0       0.00      0.00      0.00       105\n",
      "   Genome8.0       0.00      0.00      0.00       892\n",
      "   Genome9.0       0.00      0.00      0.00        67\n",
      "\n",
      "    accuracy                           0.12      9578\n",
      "   macro avg       0.01      0.04      0.01      9578\n",
      "weighted avg       0.01      0.12      0.03      9578\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/utils/validation.py:1339: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().\n",
      "  y = column_or_1d(y, warn=True)\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Classification Report untuk Support Vector Machine:\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "  Genome10.0       0.00      0.00      0.00       380\n",
      "  Genome10.1       0.24      0.66      0.36       846\n",
      "  Genome11.0       0.00      0.00      0.00         3\n",
      "  Genome12.0       0.00      0.00      0.00        59\n",
      "  Genome13.0       0.00      0.00      0.00       277\n",
      "  Genome14.0       0.00      0.00      0.00       538\n",
      "  Genome15.0       0.00      0.00      0.00       631\n",
      "  Genome16.0       0.00      0.00      0.00        54\n",
      "  Genome17.0       0.00      0.00      0.00       258\n",
      "  Genome18.0       0.00      0.00      0.00       356\n",
      "   Genome2.0       0.23      0.11      0.15       883\n",
      "  Genome20.0       0.00      0.00      0.00       110\n",
      "  Genome22.0       0.26      0.61      0.36      1142\n",
      "  Genome23.0       0.00      0.00      0.00       326\n",
      "  Genome24.0       0.00      0.00      0.00        59\n",
      "   Genome3.0       0.26      0.51      0.34      1005\n",
      "   Genome4.0       0.42      0.36      0.39       662\n",
      "   Genome4.1       0.00      0.00      0.00       180\n",
      "   Genome5.0       0.00      0.00      0.00        44\n",
      "   Genome6.0       0.26      0.45      0.33       701\n",
      "   Genome7.0       0.00      0.00      0.00       105\n",
      "   Genome8.0       0.23      0.10      0.14       892\n",
      "   Genome9.0       0.00      0.00      0.00        67\n",
      "\n",
      "    accuracy                           0.26      9578\n",
      "   macro avg       0.08      0.12      0.09      9578\n",
      "weighted avg       0.17      0.26      0.19      9578\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/neighbors/_classification.py:238: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().\n",
      "  return self._fit(X, y)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Classification Report untuk K-Nearest Neighbors:\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "  Genome10.0       0.09      0.14      0.11       380\n",
      "  Genome10.1       0.19      0.33      0.24       846\n",
      "  Genome11.0       0.00      0.00      0.00         3\n",
      "  Genome12.0       0.00      0.00      0.00        59\n",
      "  Genome13.0       0.05      0.04      0.04       277\n",
      "  Genome14.0       0.18      0.18      0.18       538\n",
      "  Genome15.0       0.12      0.15      0.13       631\n",
      "  Genome16.0       0.00      0.00      0.00        54\n",
      "  Genome17.0       0.06      0.05      0.05       258\n",
      "  Genome18.0       0.15      0.13      0.14       356\n",
      "   Genome2.0       0.18      0.18      0.18       883\n",
      "  Genome20.0       0.04      0.01      0.02       110\n",
      "  Genome22.0       0.23      0.24      0.24      1142\n",
      "  Genome23.0       0.13      0.06      0.09       326\n",
      "  Genome24.0       0.00      0.00      0.00        59\n",
      "   Genome3.0       0.25      0.26      0.25      1005\n",
      "   Genome4.0       0.31      0.33      0.32       662\n",
      "   Genome4.1       0.09      0.05      0.07       180\n",
      "   Genome5.0       0.00      0.00      0.00        44\n",
      "   Genome6.0       0.22      0.18      0.20       701\n",
      "   Genome7.0       0.00      0.00      0.00       105\n",
      "   Genome8.0       0.19      0.14      0.16       892\n",
      "   Genome9.0       0.24      0.12      0.16        67\n",
      "\n",
      "    accuracy                           0.19      9578\n",
      "   macro avg       0.12      0.11      0.11      9578\n",
      "weighted avg       0.18      0.19      0.18      9578\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model</th>\n",
       "      <th>Accuracy</th>\n",
       "      <th>Precision</th>\n",
       "      <th>Recall</th>\n",
       "      <th>F1 Score</th>\n",
       "      <th>Time (seconds)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Random Forest</td>\n",
       "      <td>0.595740</td>\n",
       "      <td>0.575911</td>\n",
       "      <td>0.595740</td>\n",
       "      <td>0.548531</td>\n",
       "      <td>40.727412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Logistic Regression</td>\n",
       "      <td>0.119232</td>\n",
       "      <td>0.014216</td>\n",
       "      <td>0.119232</td>\n",
       "      <td>0.025403</td>\n",
       "      <td>25.823937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Support Vector Machine</td>\n",
       "      <td>0.262372</td>\n",
       "      <td>0.170694</td>\n",
       "      <td>0.262372</td>\n",
       "      <td>0.188856</td>\n",
       "      <td>201.461130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>K-Nearest Neighbors</td>\n",
       "      <td>0.187617</td>\n",
       "      <td>0.181754</td>\n",
       "      <td>0.187617</td>\n",
       "      <td>0.181097</td>\n",
       "      <td>0.965518</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Model  Accuracy  Precision    Recall  F1 Score  \\\n",
       "0           Random Forest  0.595740   0.575911  0.595740  0.548531   \n",
       "1     Logistic Regression  0.119232   0.014216  0.119232  0.025403   \n",
       "2  Support Vector Machine  0.262372   0.170694  0.262372  0.188856   \n",
       "3     K-Nearest Neighbors  0.187617   0.181754  0.187617  0.181097   \n",
       "\n",
       "   Time (seconds)  \n",
       "0       40.727412  \n",
       "1       25.823937  \n",
       "2      201.461130  \n",
       "3        0.965518  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# List model yang akan digunakan\n",
    "models = {\n",
    "    \"Random Forest\": RandomForestClassifier(n_estimators=100, random_state=42),\n",
    "    \"Logistic Regression\": LogisticRegression(random_state=42),\n",
    "    \"Support Vector Machine\": SVC(random_state=42),\n",
    "    \"K-Nearest Neighbors\": KNeighborsClassifier(),\n",
    "    #\"Gradient Boosting\": GradientBoostingClassifier(n_estimators=100, random_state=42),\n",
    "    #\"MLP Classifier\": MLPClassifier(random_state=42, max_iter=500)\n",
    "}\n",
    "\n",
    "# Menyimpan hasil metrik dan waktu komputasi\n",
    "results = []\n",
    "\n",
    "# Loop melalui semua model dan evaluasi performanya\n",
    "for model_name, model in models.items():\n",
    "    start_time = time.time()  # Mulai penghitungan waktu\n",
    "    \n",
    "    # Latih model\n",
    "    model.fit(X_train, y_train)\n",
    "    \n",
    "    # Prediksi menggunakan data testing\n",
    "    y_pred = model.predict(X_test)\n",
    "    \n",
    "    # Hitung waktu komputasi\n",
    "    elapsed_time = time.time() - start_time\n",
    "    \n",
    "    # Evaluasi model dengan berbagai metrik\n",
    "    accuracy = accuracy_score(y_test, y_pred)\n",
    "    precision = precision_score(y_test, y_pred, average='weighted')\n",
    "    recall = recall_score(y_test, y_pred, average='weighted')\n",
    "    f1 = f1_score(y_test, y_pred, average='weighted')\n",
    "    \n",
    "    # Simpan hasil dalam dictionary\n",
    "    results.append({\n",
    "        \"Model\": model_name,\n",
    "        \"Accuracy\": accuracy,\n",
    "        \"Precision\": precision,\n",
    "        \"Recall\": recall,\n",
    "        \"F1 Score\": f1,\n",
    "        \"Time (seconds)\": elapsed_time\n",
    "    })\n",
    "    \n",
    "    # Tampilkan hasil classification report untuk setiap model\n",
    "    print(f\"\\nClassification Report untuk {model_name}:\")\n",
    "    print(classification_report(y_test, y_pred))\n",
    "\n",
    "# Tampilkan hasil dalam bentuk dataframe untuk mempermudah pembacaan\n",
    "import pandas as pd\n",
    "results_df = pd.DataFrame(results)\n",
    "results_df# List model yang akan digunakan\n",
    "models = {\n",
    "    \"Random Forest\": RandomForestClassifier(n_estimators=100, random_state=42),\n",
    "    \"Logistic Regression\": LogisticRegression(random_state=42),\n",
    "    \"Support Vector Machine\": SVC(random_state=42),\n",
    "    \"K-Nearest Neighbors\": KNeighborsClassifier(),\n",
    "    #\"Gradient Boosting\": GradientBoostingClassifier(n_estimators=100, random_state=42),\n",
    "    #\"MLP Classifier\": MLPClassifier(random_state=42, max_iter=500)\n",
    "}\n",
    "\n",
    "# Menyimpan hasil metrik dan waktu komputasi\n",
    "results = []\n",
    "\n",
    "# Loop melalui semua model dan evaluasi performanya\n",
    "for model_name, model in models.items():\n",
    "    start_time = time.time()  # Mulai penghitungan waktu\n",
    "    \n",
    "    # Latih model\n",
    "    model.fit(X_train, y_train)\n",
    "    \n",
    "    # Prediksi menggunakan data testing\n",
    "    y_pred = model.predict(X_test)\n",
    "    \n",
    "    # Hitung waktu komputasi\n",
    "    elapsed_time = time.time() - start_time\n",
    "    \n",
    "    # Evaluasi model dengan berbagai metrik\n",
    "    accuracy = accuracy_score(y_test, y_pred)\n",
    "    precision = precision_score(y_test, y_pred, average='weighted')\n",
    "    recall = recall_score(y_test, y_pred, average='weighted')\n",
    "    f1 = f1_score(y_test, y_pred, average='weighted')\n",
    "    \n",
    "    # Simpan hasil dalam dictionary\n",
    "    results.append({\n",
    "        \"Model\": model_name,\n",
    "        \"Accuracy\": accuracy,\n",
    "        \"Precision\": precision,\n",
    "        \"Recall\": recall,\n",
    "        \"F1 Score\": f1,\n",
    "        \"Time (seconds)\": elapsed_time\n",
    "    })\n",
    "    \n",
    "    # Tampilkan hasil classification report untuk setiap model\n",
    "    print(f\"\\nClassification Report untuk {model_name}:\")\n",
    "    print(classification_report(y_test, y_pred))\n",
    "\n",
    "# Tampilkan hasil dalam bentuk dataframe untuk mempermudah pembacaan\n",
    "import pandas as pd\n",
    "results_df = pd.DataFrame(results)\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93f32311-1b80-45e2-a60e-96a7bc7488ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# List model yang akan digunakan\n",
    "models = {\n",
    "    # \"Random Forest\": RandomForestClassifier(n_estimators=100, random_state=42),\n",
    "    # \"Logistic Regression\": LogisticRegression(random_state=42),\n",
    "    # \"Support Vector Machine\": SVC(random_state=42),\n",
    "    # \"K-Nearest Neighbors\": KNeighborsClassifier(),\n",
    "    # \"Gradient Boosting\": GradientBoostingClassifier(n_estimators=100, random_state=42),\n",
    "    \"MLP Classifier\": MLPClassifier(random_state=42, max_iter=100)\n",
    "}\n",
    "\n",
    "# Menyimpan hasil metrik dan waktu komputasi\n",
    "results = []\n",
    "\n",
    "# Loop melalui semua model dan evaluasi performanya\n",
    "for model_name, model in models.items():\n",
    "    start_time = time.time()  # Mulai penghitungan waktu\n",
    "    \n",
    "    # Latih model\n",
    "    model.fit(X_train, y_train)\n",
    "    \n",
    "    # Prediksi menggunakan data testing\n",
    "    y_pred = model.predict(X_test)\n",
    "    \n",
    "    # Hitung waktu komputasi\n",
    "    elapsed_time = time.time() - start_time\n",
    "    \n",
    "    # Evaluasi model dengan berbagai metrik\n",
    "    accuracy = accuracy_score(y_test, y_pred)\n",
    "    precision = precision_score(y_test, y_pred, average='weighted')\n",
    "    recall = recall_score(y_test, y_pred, average='weighted')\n",
    "    f1 = f1_score(y_test, y_pred, average='weighted')\n",
    "    \n",
    "    # Simpan hasil dalam dictionary\n",
    "    results.append({\n",
    "        \"Model\": model_name,\n",
    "        \"Accuracy\": accuracy,\n",
    "        \"Precision\": precision,\n",
    "        \"Recall\": recall,\n",
    "        \"F1 Score\": f1,\n",
    "        \"Time (seconds)\": elapsed_time\n",
    "    })\n",
    "    \n",
    "    # Tampilkan hasil classification report untuk setiap model\n",
    "    print(f\"\\nClassification Report untuk {model_name}:\")\n",
    "    print(classification_report(y_test, y_pred))\n",
    "\n",
    "dff = pd.DataFrame(results)\n",
    "dff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6566a66f-916c-4257-9497-ca916bd14982",
   "metadata": {},
   "outputs": [],
   "source": [
    "hasil = pd.concat([results_df, dff], ignore_index=True)\n",
    "hasil"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de576ba1-0469-44be-9597-7b34456ffce0",
   "metadata": {},
   "outputs": [],
   "source": [
    "hasil.to_csv('../../result/Kmers_camisim.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afcd99b4-a614-4b57-977f-4211ca99983c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Reads"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce8b4835-ed2e-42b7-b1e3-20db3bb78507",
   "metadata": {},
   "source": [
    "## Feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "05e558b8-1767-4d8f-87f8-babe392f81fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                     AAAA          AAAT          AAAG          AAAC  \\\n",
      "S0R0/1       1.162772e-07  1.162772e-07  1.162772e-07  1.162784e-02   \n",
      "S0R0/2       1.449247e-07  1.449247e-07  1.449247e-07  1.449247e-07   \n",
      "S0R1/1       1.351327e-07  1.351327e-07  1.351327e-07  1.351340e-02   \n",
      "S0R1/2       1.190457e-07  1.190457e-07  1.190457e-07  1.190457e-07   \n",
      "S0R2/1       1.851805e-07  1.851805e-07  1.851805e-07  1.851805e-07   \n",
      "...                   ...           ...           ...           ...   \n",
      "S0R333324/2  1.666629e-07  1.666629e-07  1.666629e-07  1.666629e-07   \n",
      "S0R333325/1  1.136346e-07  1.136346e-07  1.136346e-07  1.136346e-07   \n",
      "S0R333325/2  1.333309e-07  1.333309e-07  1.333322e-02  1.333322e-02   \n",
      "S0R333326/1  1.408424e-07  1.408424e-07  1.408438e-02  1.408424e-07   \n",
      "S0R333326/2  3.260832e-02  2.173892e-02  1.086951e-02  1.086951e-02   \n",
      "\n",
      "                     AATA          AATT          AATG          AATC  \\\n",
      "S0R0/1       1.162772e-07  1.162772e-07  1.162772e-07  1.162772e-07   \n",
      "S0R0/2       1.449247e-07  1.449247e-07  1.449247e-07  1.449247e-07   \n",
      "S0R1/1       1.351327e-07  1.351327e-07  1.351340e-02  1.351340e-02   \n",
      "S0R1/2       1.190457e-07  1.190457e-07  2.380926e-02  1.190469e-02   \n",
      "S0R2/1       1.851805e-07  1.851805e-07  1.851805e-07  1.851805e-07   \n",
      "...                   ...           ...           ...           ...   \n",
      "S0R333324/2  1.666629e-07  1.666629e-07  1.666646e-02  1.666629e-07   \n",
      "S0R333325/1  1.136346e-07  1.136346e-07  1.136346e-07  1.136346e-07   \n",
      "S0R333325/2  1.333322e-02  1.333309e-07  1.333309e-07  1.333309e-07   \n",
      "S0R333326/1  1.408424e-07  1.408438e-02  1.408424e-07  1.408424e-07   \n",
      "S0R333326/2  2.173892e-02  2.173892e-02  1.086940e-07  1.086940e-07   \n",
      "\n",
      "                     AAGA          AAGT  ...          CAAG          CATG  \\\n",
      "S0R0/1       1.162772e-07  1.162772e-07  ...  1.162784e-02  1.162772e-07   \n",
      "S0R0/2       1.449247e-07  1.449247e-07  ...  1.449247e-07  1.449247e-07   \n",
      "S0R1/1       1.351340e-02  2.702667e-02  ...  1.351340e-02  1.351340e-02   \n",
      "S0R1/2       1.190457e-07  1.190457e-07  ...  1.190457e-07  2.380926e-02   \n",
      "S0R2/1       1.851805e-07  1.851805e-07  ...  1.851805e-07  1.851824e-02   \n",
      "...                   ...           ...  ...           ...           ...   \n",
      "S0R333324/2  1.666646e-02  1.666629e-07  ...  1.666646e-02  1.666646e-02   \n",
      "S0R333325/1  1.136346e-07  1.136346e-07  ...  1.136357e-02  1.136357e-02   \n",
      "S0R333325/2  1.333309e-07  1.333322e-02  ...  1.333309e-07  3.999941e-02   \n",
      "S0R333326/1  1.408424e-07  1.408424e-07  ...  1.408438e-02  1.408438e-02   \n",
      "S0R333326/2  1.086940e-07  1.086940e-07  ...  1.086940e-07  1.086940e-07   \n",
      "\n",
      "                     CAGG          CACG          CTAG          CCAG  \\\n",
      "S0R0/1       1.162772e-07  1.162784e-02  1.162772e-07  3.488329e-02   \n",
      "S0R0/2       2.898508e-02  1.449261e-02  1.449247e-07  1.449247e-07   \n",
      "S0R1/1       1.351327e-07  1.351340e-02  1.351327e-07  1.351327e-07   \n",
      "S0R1/2       1.190469e-02  1.190457e-07  1.190457e-07  1.190469e-02   \n",
      "S0R2/1       1.851824e-02  1.851805e-07  1.851805e-07  1.851805e-07   \n",
      "...                   ...           ...           ...           ...   \n",
      "S0R333324/2  1.666629e-07  1.666629e-07  1.666629e-07  1.666629e-07   \n",
      "S0R333325/1  1.136357e-02  1.136346e-07  1.136346e-07  1.136357e-02   \n",
      "S0R333325/2  2.666632e-02  1.333322e-02  1.333309e-07  1.333309e-07   \n",
      "S0R333326/1  1.408438e-02  1.408424e-07  1.408424e-07  1.408438e-02   \n",
      "S0R333326/2  1.086951e-02  1.086940e-07  1.086951e-02  1.086951e-02   \n",
      "\n",
      "                     CGAG          CCCG          CGCG          CCGG  \n",
      "S0R0/1       2.325556e-02  1.162784e-02  3.488329e-02  3.488329e-02  \n",
      "S0R0/2       2.898508e-02  1.449261e-02  1.449247e-07  1.449247e-07  \n",
      "S0R1/1       1.351327e-07  1.351327e-07  1.351327e-07  6.756646e-02  \n",
      "S0R1/2       1.190469e-02  1.190457e-07  1.190469e-02  1.190469e-02  \n",
      "S0R2/1       1.851805e-07  1.851805e-07  7.407239e-02  1.851805e-07  \n",
      "...                   ...           ...           ...           ...  \n",
      "S0R333324/2  1.666629e-07  1.666629e-07  3.333274e-02  1.666629e-07  \n",
      "S0R333325/1  4.545396e-02  1.136346e-07  4.545396e-02  1.136346e-07  \n",
      "S0R333325/2  1.333309e-07  1.333309e-07  2.666632e-02  1.333322e-02  \n",
      "S0R333326/1  1.408424e-07  1.408438e-02  1.408424e-07  1.408438e-02  \n",
      "S0R333326/2  1.086951e-02  1.086951e-02  2.173892e-02  3.260832e-02  \n",
      "\n",
      "[666654 rows x 136 columns]\n"
     ]
    }
   ],
   "source": [
    "read_file = './reads/anonymous_reads.fq.gz'  # Updated file path to your uploaded file\n",
    "kmer_len = 4  # Original k-mer length\n",
    "length_threshold = 100  # Example sequence length threshold\n",
    "\n",
    "# Generate spaced k-mer features\n",
    "featureR_df = generate_kmer_features(read_file, length_threshold, kmer_len)\n",
    "print(featureR_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5e7e656b-6d3f-4118-8f77-f74c94c53bad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Genome19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genome19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Genome19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Genome19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Genome18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666649</th>\n",
       "      <td>Genome18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666650</th>\n",
       "      <td>Genome19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666651</th>\n",
       "      <td>Genome19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666652</th>\n",
       "      <td>Genome19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666653</th>\n",
       "      <td>Genome19.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>666654 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         genome_id\n",
       "0       Genome19.0\n",
       "1       Genome19.0\n",
       "2       Genome19.0\n",
       "3       Genome19.0\n",
       "4       Genome18.0\n",
       "...            ...\n",
       "666649  Genome18.0\n",
       "666650  Genome19.0\n",
       "666651  Genome19.0\n",
       "666652  Genome19.0\n",
       "666653  Genome19.0\n",
       "\n",
       "[666654 rows x 1 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labelR_path = \"./reads/reads_mapping.tsv.gz\"\n",
    "labelR_df = pd.read_csv(labelR_path, sep='\\t')\n",
    "labelR_df = labelR_df[[\"genome_id\"]]\n",
    "labelR_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a2b55c96-5613-4532-837b-a5a82addf271",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Jumlah data latih per kelas:\n",
      "(array(['Genome10.0', 'Genome10.1', 'Genome11.0', 'Genome12.0',\n",
      "       'Genome13.0', 'Genome14.0', 'Genome15.0', 'Genome16.0',\n",
      "       'Genome17.0', 'Genome18.0', 'Genome19.0', 'Genome2.0',\n",
      "       'Genome20.0', 'Genome22.0', 'Genome23.0', 'Genome24.0',\n",
      "       'Genome3.0', 'Genome4.0', 'Genome4.1', 'Genome5.0', 'Genome6.0',\n",
      "       'Genome7.0', 'Genome8.0', 'Genome9.0'], dtype=object), array([  2843,   9214,     37,    438,   2081,   8187,   5417,    338,\n",
      "         1936,  69103, 367831,  12720,    853,  13194,   2856,    444,\n",
      "         9619,   5141,   1501,    340,   7383,    656,  10800,    391]))\n",
      "\n",
      "Jumlah data uji per kelas:\n",
      "(array(['Genome10.0', 'Genome10.1', 'Genome11.0', 'Genome12.0',\n",
      "       'Genome13.0', 'Genome14.0', 'Genome15.0', 'Genome16.0',\n",
      "       'Genome17.0', 'Genome18.0', 'Genome19.0', 'Genome2.0',\n",
      "       'Genome20.0', 'Genome22.0', 'Genome23.0', 'Genome24.0',\n",
      "       'Genome3.0', 'Genome4.0', 'Genome4.1', 'Genome5.0', 'Genome6.0',\n",
      "       'Genome7.0', 'Genome8.0', 'Genome9.0'], dtype=object), array([  759,  2264,     9,   106,   491,  2001,  1323,    94,   446,\n",
      "       17141, 92345,  3170,   189,  3268,   730,   118,  2399,  1313,\n",
      "         369,    88,  1859,   140,  2632,    77]))\n"
     ]
    }
   ],
   "source": [
    "features = featureR_df\n",
    "labels = labelR_df\n",
    "\n",
    "# Pisahkan data menjadi training dan testing set (80% training, 20% testing)\n",
    "X_train, X_test, y_train, y_test = train_test_split(features, labels, test_size=0.2, random_state=42)\n",
    "\n",
    "print(\"Jumlah data latih per kelas:\")\n",
    "print(np.unique(y_train, return_counts=True))\n",
    "\n",
    "print(\"\\nJumlah data uji per kelas:\")\n",
    "print(np.unique(y_test, return_counts=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6790daae-39bf-49e3-8723-724c4245e1cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/sklearn/base.py:1473: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().\n",
      "  return fit_method(estimator, *args, **kwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Akurasi Model:\n",
      "0.6923746165557897\n",
      "\n",
      "Classification Report:\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "  Genome10.0       0.00      0.00      0.00       759\n",
      "  Genome10.1       0.00      0.00      0.00      2264\n",
      "  Genome11.0       0.00      0.00      0.00         9\n",
      "  Genome12.0       0.00      0.00      0.00       106\n",
      "  Genome13.0       0.00      0.00      0.00       491\n",
      "  Genome14.0       0.00      0.00      0.00      2001\n",
      "  Genome15.0       0.00      0.00      0.00      1323\n",
      "  Genome16.0       0.00      0.00      0.00        94\n",
      "  Genome17.0       0.00      0.00      0.00       446\n",
      "  Genome18.0       0.14      0.00      0.00     17141\n",
      "  Genome19.0       0.69      1.00      0.82     92345\n",
      "   Genome2.0       0.00      0.00      0.00      3170\n",
      "  Genome20.0       0.00      0.00      0.00       189\n",
      "  Genome22.0       0.00      0.00      0.00      3268\n",
      "  Genome23.0       0.00      0.00      0.00       730\n",
      "  Genome24.0       0.00      0.00      0.00       118\n",
      "   Genome3.0       0.00      0.00      0.00      2399\n",
      "   Genome4.0       0.00      0.00      0.00      1313\n",
      "   Genome4.1       0.00      0.00      0.00       369\n",
      "   Genome5.0       0.00      0.00      0.00        88\n",
      "   Genome6.0       0.00      0.00      0.00      1859\n",
      "   Genome7.0       0.00      0.00      0.00       140\n",
      "   Genome8.0       0.00      0.00      0.00      2632\n",
      "   Genome9.0       0.00      0.00      0.00        77\n",
      "\n",
      "    accuracy                           0.69    133331\n",
      "   macro avg       0.03      0.04      0.03    133331\n",
      "weighted avg       0.50      0.69      0.57    133331\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
     ]
    }
   ],
   "source": [
    "# Inisialisasi model Random Forest\n",
    "rf_model = RandomForestClassifier(n_estimators=100, random_state=42)\n",
    "\n",
    "# Latih model\n",
    "rf_model.fit(X_train, y_train)\n",
    "\n",
    "# Prediksi menggunakan data testing\n",
    "y_pred = rf_model.predict(X_test)\n",
    "\n",
    "# Tampilkan hasil akurasi dan classification report\n",
    "print(\"\\nAkurasi Model:\")\n",
    "print(accuracy_score(y_test, y_pred))\n",
    "\n",
    "print(\"\\nClassification Report:\")\n",
    "print(classification_report(y_test, y_pred))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8238cf9b-f486-4053-9b32-804b56bfd206",
   "metadata": {},
   "outputs": [],
   "source": [
    "hasil.to_csv('../../result/Kmers+abundance_camisim.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
