{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "42afb47b-86c5-465d-8127-e5ab69643813",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import gzip\n",
    "import re\n",
    "import numpy as np\n",
    "from itertools import product\n",
    "from joblib import Parallel, delayed\n",
    "from tqdm.auto import tqdm\n",
    "from collections import defaultdict, Counter, OrderedDict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df3966da-b1e6-4ce7-a493-bce87cc3441c",
   "metadata": {},
   "source": [
    "## Main Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b36aadef-6d9a-4789-914f-ed746309cd15",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to read FASTA file and yield name and sequence\n",
    "def fasta_iter(fasta_file):\n",
    "    with gzip.open(fasta_file, 'rt') as file:\n",
    "        name = None\n",
    "        seqs = []\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\">\"):\n",
    "                if name:\n",
    "                    yield name, ''.join(seqs)\n",
    "                name = line[1:]  # Skip '>'\n",
    "                seqs = []\n",
    "            else:\n",
    "                seqs.append(line)\n",
    "        if name:\n",
    "            yield name, ''.join(seqs)  # Yield the last entry\n",
    "\n",
    "# Function to read FASTQ file and yield name and sequence\n",
    "def fastq_iter(fastq_file):\n",
    "    with gzip.open(fastq_file, 'rt') as file:\n",
    "        while True:\n",
    "            name = file.readline().strip()[1:]  # Skip '@' in FASTQ\n",
    "            seq = file.readline().strip()\n",
    "            file.readline()  # Skip the '+' line\n",
    "            file.readline()  # Skip the quality line\n",
    "            if not name or not seq:\n",
    "                break\n",
    "            yield name, seq\n",
    "\n",
    "# Function to generate feature mapping for k-mers with sequences as keys\n",
    "def generate_feature_mapping(kmer_len):\n",
    "    BASE_COMPLEMENT = {\"A\": \"T\", \"T\": \"A\", \"G\": \"C\", \"C\": \"G\"}\n",
    "    kmer_hash = {}\n",
    "    kmer_names = []\n",
    "    counter = 0\n",
    "    for kmer in product(\"ATGC\", repeat=kmer_len):\n",
    "        kmer = ''.join(kmer)\n",
    "        rev_compl = ''.join([BASE_COMPLEMENT[x] for x in reversed(kmer)])  # Generate reverse complement\n",
    "        if kmer not in kmer_hash and rev_compl not in kmer_hash:\n",
    "            # Use canonical k-mer (lexicographically smaller of kmer and rev_compl)\n",
    "            canonical_kmer = min(kmer, rev_compl)\n",
    "            kmer_hash[canonical_kmer] = counter\n",
    "            kmer_names.append(canonical_kmer)  # Store the k-mer for naming features\n",
    "            counter += 1\n",
    "    return kmer_hash, kmer_names, counter\n",
    "    \n",
    "# Function to detect file type based on file extension\n",
    "def detect_file_type(file_path):\n",
    "    ext = os.path.splitext(file_path)[-1].lower()  # Get file extension and convert to lowercase\n",
    "    # Handle gzipped files\n",
    "    if ext == '.gz':\n",
    "        ext = os.path.splitext(file_path[:-3])[-1].lower()  # Check the extension before \".gz\"\n",
    "        \n",
    "    if ext in ['.fasta', '.fa', '.fna']:\n",
    "        return 'fasta'\n",
    "    elif ext in ['.fastq', '.fq']:\n",
    "        return 'fastq'\n",
    "    else:\n",
    "        raise ValueError(\"Unsupported file extension. Please provide a valid FASTA or FASTQ file.\")\n",
    "        \n",
    "# Function to generate k-mer features from FASTA or FASTQ\n",
    "def generate_kmer_features(folder_path, length_threshold, kmer_len, split=False, split_threshold=0):\n",
    "    # List all files in the folder with .fna.gz extension\n",
    "    files = [os.path.join(folder_path, f) for f in os.listdir(folder_path) if f.endswith('.fna.gz')]\n",
    "    \n",
    "    kmer_dict, kmer_names, nr_features = generate_feature_mapping(kmer_len)\n",
    "    composition = OrderedDict()\n",
    "\n",
    "    for file_path in files:\n",
    "        # Automatically detect file type from file extension\n",
    "        file_type = detect_file_type(file_path)\n",
    "\n",
    "        # Select the appropriate file iterator based on the file type\n",
    "        if file_type == 'fasta':\n",
    "            seq_iter = fasta_iter(file_path)\n",
    "        elif file_type == 'fastq':\n",
    "            seq_iter = fastq_iter(file_path)\n",
    "        else:\n",
    "            raise ValueError(\"Unsupported file type. Use 'fasta' or 'fastq'.\")\n",
    "\n",
    "        def seq_list():\n",
    "            for h, seq in seq_iter:\n",
    "                # Ambil hanya ID sebelum karakter \"|\"\n",
    "                h = h.split('|')[0].strip()\n",
    "                \n",
    "                if not split:\n",
    "                    yield h, seq\n",
    "                elif len(seq) >= split_threshold:\n",
    "                    half = len(seq) // 2\n",
    "                    yield (h + '_1', seq[:half])\n",
    "                    yield (h + '_2', seq[half:])\n",
    "\n",
    "        for h, seq in seq_list():\n",
    "            if len(seq) < length_threshold:\n",
    "                continue\n",
    "            norm_seq = str(seq).upper()\n",
    "            kmers = [kmer_dict[norm_seq[i:i+kmer_len]]\n",
    "                     for i in range(len(norm_seq) - kmer_len + 1)\n",
    "                     if norm_seq[i:i+kmer_len] in kmer_dict]  # Ignore kmers with non-canonical bases\n",
    "            composition[f\"{file_path}_{h}\"] = np.bincount(np.array(kmers, dtype=np.int64), minlength=nr_features)\n",
    "\n",
    "    # Create DataFrame with k-mer names as column headers\n",
    "    df = pd.DataFrame.from_dict(composition, orient='index', dtype=float, columns=kmer_names)\n",
    "\n",
    "    df = df.apply(lambda x: x + 1e-5)  # Smoothing\n",
    "    df = df.div(df.sum(axis=1), axis=0)  # Normalize by row (sequence)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35ad17f4-ea82-4c31-b4ef-fad1fa8a64b8",
   "metadata": {},
   "source": [
    "## Feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "149aa058-c277-4dc2-9fe1-582b4a4aba59",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAAAA</th>\n",
       "      <th>AAAAT</th>\n",
       "      <th>AAAAG</th>\n",
       "      <th>AAAAC</th>\n",
       "      <th>AAATA</th>\n",
       "      <th>AAATT</th>\n",
       "      <th>AAATG</th>\n",
       "      <th>AAATC</th>\n",
       "      <th>AAAGA</th>\n",
       "      <th>AAAGT</th>\n",
       "      <th>...</th>\n",
       "      <th>CCGAG</th>\n",
       "      <th>CGGAG</th>\n",
       "      <th>CCTCG</th>\n",
       "      <th>CGACG</th>\n",
       "      <th>CCACG</th>\n",
       "      <th>CCCCG</th>\n",
       "      <th>CGCCG</th>\n",
       "      <th>CCGCG</th>\n",
       "      <th>CCAGG</th>\n",
       "      <th>CCCGG</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1</th>\n",
       "      <td>4.716867e-08</td>\n",
       "      <td>4.716867e-08</td>\n",
       "      <td>4.716867e-08</td>\n",
       "      <td>4.716914e-03</td>\n",
       "      <td>4.716867e-08</td>\n",
       "      <td>4.716867e-08</td>\n",
       "      <td>4.716867e-08</td>\n",
       "      <td>4.716867e-08</td>\n",
       "      <td>4.716914e-03</td>\n",
       "      <td>4.716914e-03</td>\n",
       "      <td>...</td>\n",
       "      <td>4.716867e-08</td>\n",
       "      <td>9.433782e-03</td>\n",
       "      <td>4.716867e-08</td>\n",
       "      <td>4.716914e-03</td>\n",
       "      <td>4.716867e-08</td>\n",
       "      <td>4.716914e-03</td>\n",
       "      <td>0.014151</td>\n",
       "      <td>4.716867e-08</td>\n",
       "      <td>4.716867e-08</td>\n",
       "      <td>1.415065e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1</th>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936968e-03</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936968e-03</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>3.936968e-03</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936968e-03</td>\n",
       "      <td>3.936968e-03</td>\n",
       "      <td>0.031495</td>\n",
       "      <td>3.936968e-03</td>\n",
       "      <td>7.873896e-03</td>\n",
       "      <td>3.936968e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1</th>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936968e-03</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936968e-03</td>\n",
       "      <td>3.936968e-03</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>0.003937</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>7.873896e-03</td>\n",
       "      <td>3.936929e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1</th>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>3.802245e-03</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>2.281328e-02</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>7.604453e-03</td>\n",
       "      <td>0.015209</td>\n",
       "      <td>3.802245e-03</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>3.802245e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1</th>\n",
       "      <td>3.999918e-08</td>\n",
       "      <td>3.999918e-08</td>\n",
       "      <td>3.999918e-08</td>\n",
       "      <td>3.999918e-08</td>\n",
       "      <td>3.999918e-08</td>\n",
       "      <td>3.999918e-08</td>\n",
       "      <td>3.999918e-08</td>\n",
       "      <td>3.999918e-08</td>\n",
       "      <td>3.999958e-03</td>\n",
       "      <td>3.999918e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>3.999918e-08</td>\n",
       "      <td>3.999918e-08</td>\n",
       "      <td>7.999876e-03</td>\n",
       "      <td>3.999958e-03</td>\n",
       "      <td>3.999918e-08</td>\n",
       "      <td>7.999876e-03</td>\n",
       "      <td>0.012000</td>\n",
       "      <td>7.999876e-03</td>\n",
       "      <td>1.599971e-02</td>\n",
       "      <td>3.999958e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1</th>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>4.237239e-03</td>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>4.237239e-03</td>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>4.237239e-03</td>\n",
       "      <td>8.474435e-03</td>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>0.004237</td>\n",
       "      <td>4.237239e-03</td>\n",
       "      <td>4.237239e-03</td>\n",
       "      <td>4.237196e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1</th>\n",
       "      <td>4.048499e-08</td>\n",
       "      <td>4.048540e-03</td>\n",
       "      <td>4.048540e-03</td>\n",
       "      <td>4.048499e-08</td>\n",
       "      <td>4.048499e-08</td>\n",
       "      <td>4.048499e-08</td>\n",
       "      <td>4.048540e-03</td>\n",
       "      <td>4.048499e-08</td>\n",
       "      <td>4.048499e-08</td>\n",
       "      <td>4.048540e-03</td>\n",
       "      <td>...</td>\n",
       "      <td>4.048499e-08</td>\n",
       "      <td>4.048540e-03</td>\n",
       "      <td>4.048499e-08</td>\n",
       "      <td>4.048499e-08</td>\n",
       "      <td>4.048540e-03</td>\n",
       "      <td>4.048499e-08</td>\n",
       "      <td>0.028340</td>\n",
       "      <td>4.048540e-03</td>\n",
       "      <td>4.048499e-08</td>\n",
       "      <td>4.048499e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1</th>\n",
       "      <td>1.090892e-02</td>\n",
       "      <td>1.090892e-02</td>\n",
       "      <td>1.090892e-02</td>\n",
       "      <td>3.636296e-08</td>\n",
       "      <td>3.636296e-08</td>\n",
       "      <td>7.272628e-03</td>\n",
       "      <td>3.636332e-03</td>\n",
       "      <td>7.272628e-03</td>\n",
       "      <td>7.272628e-03</td>\n",
       "      <td>3.636296e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>3.636296e-08</td>\n",
       "      <td>3.636332e-03</td>\n",
       "      <td>3.636332e-03</td>\n",
       "      <td>3.636296e-08</td>\n",
       "      <td>3.636296e-08</td>\n",
       "      <td>3.636296e-08</td>\n",
       "      <td>0.003636</td>\n",
       "      <td>3.636332e-03</td>\n",
       "      <td>3.636332e-03</td>\n",
       "      <td>3.636296e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1</th>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>3.891012e-03</td>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>7.781985e-03</td>\n",
       "      <td>3.891012e-03</td>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>7.781985e-03</td>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>3.891012e-03</td>\n",
       "      <td>0.007782</td>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>3.891012e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1</th>\n",
       "      <td>3.689967e-08</td>\n",
       "      <td>3.690004e-03</td>\n",
       "      <td>3.689967e-08</td>\n",
       "      <td>3.690004e-03</td>\n",
       "      <td>1.106994e-02</td>\n",
       "      <td>3.689967e-08</td>\n",
       "      <td>3.689967e-08</td>\n",
       "      <td>3.690004e-03</td>\n",
       "      <td>3.690004e-03</td>\n",
       "      <td>3.689967e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>3.689967e-08</td>\n",
       "      <td>3.689967e-08</td>\n",
       "      <td>3.690004e-03</td>\n",
       "      <td>1.106994e-02</td>\n",
       "      <td>3.689967e-08</td>\n",
       "      <td>3.690004e-03</td>\n",
       "      <td>0.011070</td>\n",
       "      <td>3.690004e-03</td>\n",
       "      <td>3.690004e-03</td>\n",
       "      <td>7.379971e-03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100000 rows × 512 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                       AAAAA         AAAAT  \\\n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1      4.716867e-08  4.716867e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1      3.936929e-08  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1      3.936929e-08  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1      3.802207e-08  3.802207e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1      3.999918e-08  3.999918e-08   \n",
       "...                                                      ...           ...   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1   4.237196e-08  4.237196e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1   4.048499e-08  4.048540e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1   1.090892e-02  1.090892e-02   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1   3.890973e-08  3.891012e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1  3.689967e-08  3.690004e-03   \n",
       "\n",
       "                                                       AAAAG         AAAAC  \\\n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1      4.716867e-08  4.716914e-03   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1      3.936968e-03  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1      3.936929e-08  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1      3.802207e-08  3.802207e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1      3.999918e-08  3.999918e-08   \n",
       "...                                                      ...           ...   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1   4.237196e-08  4.237239e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1   4.048540e-03  4.048499e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1   1.090892e-02  3.636296e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1   3.890973e-08  3.890973e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1  3.689967e-08  3.690004e-03   \n",
       "\n",
       "                                                       AAATA         AAATT  \\\n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1      4.716867e-08  4.716867e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1      3.936929e-08  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1      3.936929e-08  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1      3.802207e-08  3.802207e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1      3.999918e-08  3.999918e-08   \n",
       "...                                                      ...           ...   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1   4.237196e-08  4.237239e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1   4.048499e-08  4.048499e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1   3.636296e-08  7.272628e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1   7.781985e-03  3.891012e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1  1.106994e-02  3.689967e-08   \n",
       "\n",
       "                                                       AAATG         AAATC  \\\n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1      4.716867e-08  4.716867e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1      3.936929e-08  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1      3.936929e-08  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1      3.802207e-08  3.802207e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1      3.999918e-08  3.999918e-08   \n",
       "...                                                      ...           ...   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1   4.237196e-08  4.237239e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1   4.048540e-03  4.048499e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1   3.636332e-03  7.272628e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1   3.890973e-08  3.890973e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1  3.689967e-08  3.690004e-03   \n",
       "\n",
       "                                                       AAAGA         AAAGT  \\\n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1      4.716914e-03  4.716914e-03   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1      3.936968e-03  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1      3.936968e-03  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1      3.802207e-08  3.802207e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1      3.999958e-03  3.999918e-08   \n",
       "...                                                      ...           ...   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1   8.474435e-03  4.237196e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1   4.048499e-08  4.048540e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1   7.272628e-03  3.636296e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1   3.890973e-08  3.890973e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1  3.690004e-03  3.689967e-08   \n",
       "\n",
       "                                                ...         CCGAG  \\\n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1      ...  4.716867e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1      ...  3.936968e-03   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1      ...  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1      ...  3.802245e-03   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1      ...  3.999918e-08   \n",
       "...                                             ...           ...   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1   ...  4.237196e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1   ...  4.048499e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1   ...  3.636296e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1   ...  3.890973e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1  ...  3.689967e-08   \n",
       "\n",
       "                                                       CGGAG         CCTCG  \\\n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1      9.433782e-03  4.716867e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1      3.936929e-08  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1      3.936929e-08  3.936968e-03   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1      3.802207e-08  3.802207e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1      3.999918e-08  7.999876e-03   \n",
       "...                                                      ...           ...   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1   4.237196e-08  4.237196e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1   4.048540e-03  4.048499e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1   3.636332e-03  3.636332e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1   7.781985e-03  3.890973e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1  3.689967e-08  3.690004e-03   \n",
       "\n",
       "                                                       CGACG         CCACG  \\\n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1      4.716914e-03  4.716867e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1      3.936929e-08  3.936968e-03   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1      3.936968e-03  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1      2.281328e-02  3.802207e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1      3.999958e-03  3.999918e-08   \n",
       "...                                                      ...           ...   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1   4.237196e-08  4.237196e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1   4.048499e-08  4.048540e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1   3.636296e-08  3.636296e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1   3.890973e-08  3.890973e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1  1.106994e-02  3.689967e-08   \n",
       "\n",
       "                                                       CCCCG     CGCCG  \\\n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1      4.716914e-03  0.014151   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1      3.936968e-03  0.031495   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1      3.936929e-08  0.003937   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1      7.604453e-03  0.015209   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1      7.999876e-03  0.012000   \n",
       "...                                                      ...       ...   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1   4.237196e-08  0.004237   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1   4.048499e-08  0.028340   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1   3.636296e-08  0.003636   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1   3.891012e-03  0.007782   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1  3.690004e-03  0.011070   \n",
       "\n",
       "                                                       CCGCG         CCAGG  \\\n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1      4.716867e-08  4.716867e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1      3.936968e-03  7.873896e-03   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1      3.936929e-08  7.873896e-03   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1      3.802245e-03  3.802207e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1      7.999876e-03  1.599971e-02   \n",
       "...                                                      ...           ...   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1   4.237239e-03  4.237239e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1   4.048540e-03  4.048499e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1   3.636332e-03  3.636332e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1   3.890973e-08  3.890973e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1  3.690004e-03  3.690004e-03   \n",
       "\n",
       "                                                       CCCGG  \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1      1.415065e-02  \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1      3.936968e-03  \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1      3.936929e-08  \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1      3.802245e-03  \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1      3.999958e-03  \n",
       "...                                                      ...  \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1   4.237196e-08  \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1   4.048499e-08  \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1   3.636296e-08  \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1   3.891012e-03  \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1  7.379971e-03  \n",
       "\n",
       "[100000 rows x 512 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.chdir(\"../../data/10famsim\") # Directory to your data folder\n",
    "fasta_folder = './'  # Updated folder path to your uploaded file\n",
    "kmer_len = 5  # Original k-mer length\n",
    "length_threshold = 100  # Example sequence length threshold\n",
    "\n",
    "# Generate spaced k-mer features\n",
    "kmer_df = generate_kmer_features(fasta_folder, length_threshold, kmer_len)\n",
    "kmer_df = pd.DataFrame(kmer_df)\n",
    "kmer_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b030146f-3502-49d6-ad5a-1e5170395687",
   "metadata": {},
   "source": [
    "## Label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ccc77d20-9b15-4645-9fb7-887c1bfa9ebc",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_006177.1 Symbiobacterium thermophilum IAM 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_006177.1 Symbiobacterium thermophilum IAM 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_006177.1 Symbiobacterium thermophilum IAM 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_006177.1 Symbiobacterium thermophilum IAM 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_006177.1 Symbiobacterium thermophilum IAM 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99995</th>\n",
       "      <td>NC_007645.1 Hahella chejuensis KCTC 2396, comp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99996</th>\n",
       "      <td>NC_007645.1 Hahella chejuensis KCTC 2396, comp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99997</th>\n",
       "      <td>NC_007645.1 Hahella chejuensis KCTC 2396, comp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99998</th>\n",
       "      <td>NC_007645.1 Hahella chejuensis KCTC 2396, comp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99999</th>\n",
       "      <td>NC_007645.1 Hahella chejuensis KCTC 2396, comp...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100000 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   Label\n",
       "0      NC_006177.1 Symbiobacterium thermophilum IAM 1...\n",
       "1      NC_006177.1 Symbiobacterium thermophilum IAM 1...\n",
       "2      NC_006177.1 Symbiobacterium thermophilum IAM 1...\n",
       "3      NC_006177.1 Symbiobacterium thermophilum IAM 1...\n",
       "4      NC_006177.1 Symbiobacterium thermophilum IAM 1...\n",
       "...                                                  ...\n",
       "99995  NC_007645.1 Hahella chejuensis KCTC 2396, comp...\n",
       "99996  NC_007645.1 Hahella chejuensis KCTC 2396, comp...\n",
       "99997  NC_007645.1 Hahella chejuensis KCTC 2396, comp...\n",
       "99998  NC_007645.1 Hahella chejuensis KCTC 2396, comp...\n",
       "99999  NC_007645.1 Hahella chejuensis KCTC 2396, comp...\n",
       "\n",
       "[100000 rows x 1 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def generate_label_dataframe(file_path):\n",
    "    labels = OrderedDict()\n",
    "    \n",
    "    # Automatically detect file type\n",
    "    file_type = detect_file_type(file_path)\n",
    "    if file_type == 'fasta':\n",
    "        seq_iter = fasta_iter(file_path)\n",
    "    elif file_type == 'fastq':\n",
    "        seq_iter = fastq_iter(file_path)\n",
    "    else:\n",
    "        raise ValueError(\"Unsupported file type. Use 'fasta' or 'fastq'.\")\n",
    "\n",
    "    for h, seq in seq_iter:\n",
    "        # Ambil hanya ID sebelum karakter \"|\"\n",
    "        seq_id = h.split('|')[0].strip()\n",
    "        # Ekstrak label di antara tanda petik dua (\"\")\n",
    "        label_match = re.search(r'\"(.*?)\"', h)\n",
    "        label = label_match.group(1) if label_match else \"Unknown\"\n",
    "        \n",
    "        # Simpan ID dan label\n",
    "        labels[seq_id] = label\n",
    "\n",
    "    # Buat DataFrame dari dictionary labels\n",
    "    label_df = pd.DataFrame.from_dict(labels, orient='index', columns=['Label'])\n",
    "    \n",
    "    return label_df\n",
    "\n",
    "folder_path = './'\n",
    "files = [os.path.join(folder_path, f) for f in os.listdir(folder_path) if f.endswith('.fna.gz')]\n",
    "\n",
    "# Menggabungkan semua label dari beberapa file\n",
    "label_df = pd.concat([generate_label_dataframe(f) for f in files])\n",
    "\n",
    "# Reset indeks jika diperlukan\n",
    "label_df.reset_index(inplace=True)\n",
    "label_df.columns = ['Sequence_ID', 'Label']\n",
    "label_df =  label_df[['Label']]\n",
    "label_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48b55240-18e9-4edd-bf34-8755a1827898",
   "metadata": {},
   "source": [
    "## Feature Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2e12dff7-c5fe-4b9c-8a3e-7240c214b8d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Statistik Deskriptif Data Fitur:\n",
      "              AAAAA         AAAAT         AAAAG         AAAAC         AAATA  \\\n",
      "count  1.000000e+05  1.000000e+05  1.000000e+05  1.000000e+05  1.000000e+05   \n",
      "mean   7.192815e-03  5.690635e-03  4.235023e-03  3.654365e-03  4.729090e-03   \n",
      "std    8.771448e-03  6.020118e-03  4.864990e-03  4.166099e-03  5.601509e-03   \n",
      "min    2.481358e-08  2.358462e-08  2.358462e-08  2.577286e-08  2.358462e-08   \n",
      "25%    4.328908e-08  4.405187e-08  4.255226e-08  4.219318e-08  4.201590e-08   \n",
      "50%    4.149331e-03  4.166619e-03  3.787843e-03  3.662972e-03  3.846117e-03   \n",
      "75%    1.153827e-02  8.695502e-03  7.352840e-03  5.434686e-03  7.812383e-03   \n",
      "max    1.051199e-01  4.498194e-02  4.297934e-02  3.401305e-02  5.395588e-02   \n",
      "\n",
      "              AAATT         AAATG         AAATC         AAAGA         AAAGT  \\\n",
      "count  1.000000e+05  1.000000e+05  1.000000e+05  1.000000e+05  1.000000e+05   \n",
      "mean   4.254107e-03  3.268449e-03  3.496504e-03  3.605722e-03  2.634730e-03   \n",
      "std    5.126718e-03  4.091042e-03  4.054192e-03  4.552977e-03  3.570478e-03   \n",
      "min    2.358462e-08  2.358462e-08  2.358462e-08  2.358462e-08  2.358462e-08   \n",
      "25%    4.149289e-08  4.098275e-08  4.149289e-08  4.166578e-08  4.015982e-08   \n",
      "50%    3.717439e-03  3.257307e-03  3.597092e-03  3.436400e-03  4.784572e-08   \n",
      "75%    7.352840e-03  4.587094e-03  4.807622e-03  4.999922e-03  4.219360e-03   \n",
      "max    4.310254e-02  3.383397e-02  4.532515e-02  5.444049e-02  3.076866e-02   \n",
      "\n",
      "       ...         CCGAG         CGGAG         CCTCG         CGACG  \\\n",
      "count  ...  1.000000e+05  1.000000e+05  1.000000e+05  1.000000e+05   \n",
      "mean   ...  1.200443e-03  1.476732e-03  1.260728e-03  1.468669e-03   \n",
      "std    ...  2.530662e-03  2.879007e-03  2.806659e-03  3.062545e-03   \n",
      "min    ...  2.577286e-08  2.481358e-08  2.358462e-08  2.577286e-08   \n",
      "25%    ...  3.846078e-08  3.860928e-08  3.831342e-08  3.860928e-08   \n",
      "50%    ...  4.237196e-08  4.273411e-08  4.201590e-08  4.255226e-08   \n",
      "75%    ...  5.263016e-08  3.496476e-03  5.181210e-08  6.493291e-08   \n",
      "max    ...  3.584167e-02  6.856948e-02  6.382866e-02  8.558366e-02   \n",
      "\n",
      "              CCACG         CCCCG         CGCCG         CCGCG         CCAGG  \\\n",
      "count  1.000000e+05  1.000000e+05  1.000000e+05  1.000000e+05  1.000000e+05   \n",
      "mean   1.297330e-03  1.542716e-03  3.374769e-03  1.743653e-03  2.283564e-03   \n",
      "std    2.706591e-03  3.127012e-03  5.871782e-03  3.477454e-03  3.966292e-03   \n",
      "min    2.358462e-08  2.358462e-08  2.577286e-08  2.358462e-08  2.577286e-08   \n",
      "25%    3.846078e-08  3.860928e-08  3.936929e-08  3.860928e-08  3.921490e-08   \n",
      "50%    4.237196e-08  4.273411e-08  4.504401e-08  4.291751e-08  4.444343e-08   \n",
      "75%    5.555398e-08  3.436400e-03  4.310293e-03  3.623157e-03  3.968213e-03   \n",
      "max    3.964673e-02  5.485118e-02  7.692097e-02  3.718934e-02  3.813481e-02   \n",
      "\n",
      "              CCCGG  \n",
      "count  1.000000e+05  \n",
      "mean   2.462758e-03  \n",
      "std    4.561096e-03  \n",
      "min    2.358462e-08  \n",
      "25%    3.890973e-08  \n",
      "50%    4.405187e-08  \n",
      "75%    3.984022e-03  \n",
      "max    1.101648e-01  \n",
      "\n",
      "[8 rows x 512 columns]\n",
      "\n",
      "Distribusi Kelas pada Data Label:\n",
      "Label                                                                                  \n",
      "NC_000964.3 Bacillus subtilis subsp. subtilis str. 168 complete genome                     10000\n",
      "NC_003910.7 Colwellia psychrerythraea 34H, complete sequence                               10000\n",
      "NC_006177.1 Symbiobacterium thermophilum IAM 14863, complete sequence                      10000\n",
      "NC_007645.1 Hahella chejuensis KCTC 2396, complete sequence                                10000\n",
      "NC_007795.1 Staphylococcus aureus subsp. aureus NCTC 8325 chromosome, complete genome      10000\n",
      "NC_008346.1 Syntrophomonas wolfei subsp. wolfei str. Goettingen G311, complete sequence    10000\n",
      "NC_008709.1 Psychromonas ingrahamii 37, complete sequence                                  10000\n",
      "NC_009922.1 Alkaliphilus oremlandii OhILAs, complete sequence                              10000\n",
      "NC_011830.1 Desulfitobacterium hafniense DCB-2, complete sequence                          10000\n",
      "NC_013422.1 Halothiobacillus neapolitanus c2, complete sequence                            10000\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Tampilkan statistik deskriptif dari data fitur\n",
    "print(\"Statistik Deskriptif Data Fitur:\")\n",
    "print(kmer_df.describe())\n",
    "\n",
    "# Tampilkan distribusi kelas dari data label\n",
    "print(\"\\nDistribusi Kelas pada Data Label:\")\n",
    "print(label_df.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "05e558b8-1767-4d8f-87f8-babe392f81fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                     AAAA          AAAT          AAAG          AAAC  \\\n",
      "S0R0/1       1.162772e-07  1.162772e-07  1.162772e-07  1.162784e-02   \n",
      "S0R0/2       1.449247e-07  1.449247e-07  1.449247e-07  1.449247e-07   \n",
      "S0R1/1       1.351327e-07  1.351327e-07  1.351327e-07  1.351340e-02   \n",
      "S0R1/2       1.190457e-07  1.190457e-07  1.190457e-07  1.190457e-07   \n",
      "S0R2/1       1.851805e-07  1.851805e-07  1.851805e-07  1.851805e-07   \n",
      "...                   ...           ...           ...           ...   \n",
      "S0R333324/2  1.666629e-07  1.666629e-07  1.666629e-07  1.666629e-07   \n",
      "S0R333325/1  1.136346e-07  1.136346e-07  1.136346e-07  1.136346e-07   \n",
      "S0R333325/2  1.333309e-07  1.333309e-07  1.333322e-02  1.333322e-02   \n",
      "S0R333326/1  1.408424e-07  1.408424e-07  1.408438e-02  1.408424e-07   \n",
      "S0R333326/2  3.260832e-02  2.173892e-02  1.086951e-02  1.086951e-02   \n",
      "\n",
      "                     AATA          AATT          AATG          AATC  \\\n",
      "S0R0/1       1.162772e-07  1.162772e-07  1.162772e-07  1.162772e-07   \n",
      "S0R0/2       1.449247e-07  1.449247e-07  1.449247e-07  1.449247e-07   \n",
      "S0R1/1       1.351327e-07  1.351327e-07  1.351340e-02  1.351340e-02   \n",
      "S0R1/2       1.190457e-07  1.190457e-07  2.380926e-02  1.190469e-02   \n",
      "S0R2/1       1.851805e-07  1.851805e-07  1.851805e-07  1.851805e-07   \n",
      "...                   ...           ...           ...           ...   \n",
      "S0R333324/2  1.666629e-07  1.666629e-07  1.666646e-02  1.666629e-07   \n",
      "S0R333325/1  1.136346e-07  1.136346e-07  1.136346e-07  1.136346e-07   \n",
      "S0R333325/2  1.333322e-02  1.333309e-07  1.333309e-07  1.333309e-07   \n",
      "S0R333326/1  1.408424e-07  1.408438e-02  1.408424e-07  1.408424e-07   \n",
      "S0R333326/2  2.173892e-02  2.173892e-02  1.086940e-07  1.086940e-07   \n",
      "\n",
      "                     AAGA          AAGT  ...          CAAG          CATG  \\\n",
      "S0R0/1       1.162772e-07  1.162772e-07  ...  1.162784e-02  1.162772e-07   \n",
      "S0R0/2       1.449247e-07  1.449247e-07  ...  1.449247e-07  1.449247e-07   \n",
      "S0R1/1       1.351340e-02  2.702667e-02  ...  1.351340e-02  1.351340e-02   \n",
      "S0R1/2       1.190457e-07  1.190457e-07  ...  1.190457e-07  2.380926e-02   \n",
      "S0R2/1       1.851805e-07  1.851805e-07  ...  1.851805e-07  1.851824e-02   \n",
      "...                   ...           ...  ...           ...           ...   \n",
      "S0R333324/2  1.666646e-02  1.666629e-07  ...  1.666646e-02  1.666646e-02   \n",
      "S0R333325/1  1.136346e-07  1.136346e-07  ...  1.136357e-02  1.136357e-02   \n",
      "S0R333325/2  1.333309e-07  1.333322e-02  ...  1.333309e-07  3.999941e-02   \n",
      "S0R333326/1  1.408424e-07  1.408424e-07  ...  1.408438e-02  1.408438e-02   \n",
      "S0R333326/2  1.086940e-07  1.086940e-07  ...  1.086940e-07  1.086940e-07   \n",
      "\n",
      "                     CAGG          CACG          CTAG          CCAG  \\\n",
      "S0R0/1       1.162772e-07  1.162784e-02  1.162772e-07  3.488329e-02   \n",
      "S0R0/2       2.898508e-02  1.449261e-02  1.449247e-07  1.449247e-07   \n",
      "S0R1/1       1.351327e-07  1.351340e-02  1.351327e-07  1.351327e-07   \n",
      "S0R1/2       1.190469e-02  1.190457e-07  1.190457e-07  1.190469e-02   \n",
      "S0R2/1       1.851824e-02  1.851805e-07  1.851805e-07  1.851805e-07   \n",
      "...                   ...           ...           ...           ...   \n",
      "S0R333324/2  1.666629e-07  1.666629e-07  1.666629e-07  1.666629e-07   \n",
      "S0R333325/1  1.136357e-02  1.136346e-07  1.136346e-07  1.136357e-02   \n",
      "S0R333325/2  2.666632e-02  1.333322e-02  1.333309e-07  1.333309e-07   \n",
      "S0R333326/1  1.408438e-02  1.408424e-07  1.408424e-07  1.408438e-02   \n",
      "S0R333326/2  1.086951e-02  1.086940e-07  1.086951e-02  1.086951e-02   \n",
      "\n",
      "                     CGAG          CCCG          CGCG          CCGG  \n",
      "S0R0/1       2.325556e-02  1.162784e-02  3.488329e-02  3.488329e-02  \n",
      "S0R0/2       2.898508e-02  1.449261e-02  1.449247e-07  1.449247e-07  \n",
      "S0R1/1       1.351327e-07  1.351327e-07  1.351327e-07  6.756646e-02  \n",
      "S0R1/2       1.190469e-02  1.190457e-07  1.190469e-02  1.190469e-02  \n",
      "S0R2/1       1.851805e-07  1.851805e-07  7.407239e-02  1.851805e-07  \n",
      "...                   ...           ...           ...           ...  \n",
      "S0R333324/2  1.666629e-07  1.666629e-07  3.333274e-02  1.666629e-07  \n",
      "S0R333325/1  4.545396e-02  1.136346e-07  4.545396e-02  1.136346e-07  \n",
      "S0R333325/2  1.333309e-07  1.333309e-07  2.666632e-02  1.333322e-02  \n",
      "S0R333326/1  1.408424e-07  1.408438e-02  1.408424e-07  1.408438e-02  \n",
      "S0R333326/2  1.086951e-02  1.086951e-02  2.173892e-02  3.260832e-02  \n",
      "\n",
      "[666654 rows x 136 columns]\n"
     ]
    }
   ],
   "source": [
    "read_file = './reads/anonymous_reads.fq.gz'  # Updated file path to your uploaded file\n",
    "kmer_len = 4  # Original k-mer length\n",
    "length_threshold = 100  # Example sequence length threshold\n",
    "\n",
    "# Generate spaced k-mer features\n",
    "featureR_df = generate_kmer_features(read_file, length_threshold, kmer_len)\n",
    "print(featureR_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5e7e656b-6d3f-4118-8f77-f74c94c53bad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Genome19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Genome19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Genome19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Genome19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Genome18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666649</th>\n",
       "      <td>Genome18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666650</th>\n",
       "      <td>Genome19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666651</th>\n",
       "      <td>Genome19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666652</th>\n",
       "      <td>Genome19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>666653</th>\n",
       "      <td>Genome19.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>666654 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         genome_id\n",
       "0       Genome19.0\n",
       "1       Genome19.0\n",
       "2       Genome19.0\n",
       "3       Genome19.0\n",
       "4       Genome18.0\n",
       "...            ...\n",
       "666649  Genome18.0\n",
       "666650  Genome19.0\n",
       "666651  Genome19.0\n",
       "666652  Genome19.0\n",
       "666653  Genome19.0\n",
       "\n",
       "[666654 rows x 1 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labelR_path = \"./reads/reads_mapping.tsv.gz\"\n",
    "labelR_df = pd.read_csv(labelR_path, sep='\\t')\n",
    "labelR_df = labelR_df[[\"genome_id\"]]\n",
    "labelR_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a2b55c96-5613-4532-837b-a5a82addf271",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Jumlah data latih per kelas:\n",
      "(array(['Genome10.0', 'Genome10.1', 'Genome11.0', 'Genome12.0',\n",
      "       'Genome13.0', 'Genome14.0', 'Genome15.0', 'Genome16.0',\n",
      "       'Genome17.0', 'Genome18.0', 'Genome19.0', 'Genome2.0',\n",
      "       'Genome20.0', 'Genome22.0', 'Genome23.0', 'Genome24.0',\n",
      "       'Genome3.0', 'Genome4.0', 'Genome4.1', 'Genome5.0', 'Genome6.0',\n",
      "       'Genome7.0', 'Genome8.0', 'Genome9.0'], dtype=object), array([  2843,   9214,     37,    438,   2081,   8187,   5417,    338,\n",
      "         1936,  69103, 367831,  12720,    853,  13194,   2856,    444,\n",
      "         9619,   5141,   1501,    340,   7383,    656,  10800,    391]))\n",
      "\n",
      "Jumlah data uji per kelas:\n",
      "(array(['Genome10.0', 'Genome10.1', 'Genome11.0', 'Genome12.0',\n",
      "       'Genome13.0', 'Genome14.0', 'Genome15.0', 'Genome16.0',\n",
      "       'Genome17.0', 'Genome18.0', 'Genome19.0', 'Genome2.0',\n",
      "       'Genome20.0', 'Genome22.0', 'Genome23.0', 'Genome24.0',\n",
      "       'Genome3.0', 'Genome4.0', 'Genome4.1', 'Genome5.0', 'Genome6.0',\n",
      "       'Genome7.0', 'Genome8.0', 'Genome9.0'], dtype=object), array([  759,  2264,     9,   106,   491,  2001,  1323,    94,   446,\n",
      "       17141, 92345,  3170,   189,  3268,   730,   118,  2399,  1313,\n",
      "         369,    88,  1859,   140,  2632,    77]))\n"
     ]
    }
   ],
   "source": [
    "features = featureR_df\n",
    "labels = labelR_df\n",
    "\n",
    "# Pisahkan data menjadi training dan testing set (80% training, 20% testing)\n",
    "X_train, X_test, y_train, y_test = train_test_split(features, labels, test_size=0.2, random_state=42)\n",
    "\n",
    "print(\"Jumlah data latih per kelas:\")\n",
    "print(np.unique(y_train, return_counts=True))\n",
    "\n",
    "print(\"\\nJumlah data uji per kelas:\")\n",
    "print(np.unique(y_test, return_counts=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6790daae-39bf-49e3-8723-724c4245e1cd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/sklearn/base.py:1473: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().\n",
      "  return fit_method(estimator, *args, **kwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Akurasi Model:\n",
      "0.6923746165557897\n",
      "\n",
      "Classification Report:\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              precision    recall  f1-score   support\n",
      "\n",
      "  Genome10.0       0.00      0.00      0.00       759\n",
      "  Genome10.1       0.00      0.00      0.00      2264\n",
      "  Genome11.0       0.00      0.00      0.00         9\n",
      "  Genome12.0       0.00      0.00      0.00       106\n",
      "  Genome13.0       0.00      0.00      0.00       491\n",
      "  Genome14.0       0.00      0.00      0.00      2001\n",
      "  Genome15.0       0.00      0.00      0.00      1323\n",
      "  Genome16.0       0.00      0.00      0.00        94\n",
      "  Genome17.0       0.00      0.00      0.00       446\n",
      "  Genome18.0       0.14      0.00      0.00     17141\n",
      "  Genome19.0       0.69      1.00      0.82     92345\n",
      "   Genome2.0       0.00      0.00      0.00      3170\n",
      "  Genome20.0       0.00      0.00      0.00       189\n",
      "  Genome22.0       0.00      0.00      0.00      3268\n",
      "  Genome23.0       0.00      0.00      0.00       730\n",
      "  Genome24.0       0.00      0.00      0.00       118\n",
      "   Genome3.0       0.00      0.00      0.00      2399\n",
      "   Genome4.0       0.00      0.00      0.00      1313\n",
      "   Genome4.1       0.00      0.00      0.00       369\n",
      "   Genome5.0       0.00      0.00      0.00        88\n",
      "   Genome6.0       0.00      0.00      0.00      1859\n",
      "   Genome7.0       0.00      0.00      0.00       140\n",
      "   Genome8.0       0.00      0.00      0.00      2632\n",
      "   Genome9.0       0.00      0.00      0.00        77\n",
      "\n",
      "    accuracy                           0.69    133331\n",
      "   macro avg       0.03      0.04      0.03    133331\n",
      "weighted avg       0.50      0.69      0.57    133331\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
     ]
    }
   ],
   "source": [
    "# Inisialisasi model Random Forest\n",
    "rf_model = RandomForestClassifier(n_estimators=100, random_state=42)\n",
    "\n",
    "# Latih model\n",
    "rf_model.fit(X_train, y_train)\n",
    "\n",
    "# Prediksi menggunakan data testing\n",
    "y_pred = rf_model.predict(X_test)\n",
    "\n",
    "# Tampilkan hasil akurasi dan classification report\n",
    "print(\"\\nAkurasi Model:\")\n",
    "print(accuracy_score(y_test, y_pred))\n",
    "\n",
    "print(\"\\nClassification Report:\")\n",
    "print(classification_report(y_test, y_pred))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8238cf9b-f486-4053-9b32-804b56bfd206",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
