{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "42afb47b-86c5-465d-8127-e5ab69643813",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import gzip\n",
    "import numpy as np\n",
    "from itertools import product\n",
    "from joblib import Parallel, delayed\n",
    "from tqdm.auto import tqdm\n",
    "from collections import defaultdict, Counter, OrderedDict\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ce099307-cb36-489b-8c11-25aeaef71788",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"../../data/10famsim\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1bef3d99-77ec-4a54-8890-1e7de6d4e90e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/jovyan/work/RCSparK/data/10famsim'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f9c23b9b-2e30-43b5-879b-03b294860855",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to read FASTA file and yield name and sequence\n",
    "def fasta_iter(fasta_file):\n",
    "    with gzip.open(fasta_file, 'rt') as file:\n",
    "        name = None\n",
    "        seqs = []\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if line.startswith(\">\"):\n",
    "                if name:\n",
    "                    yield name, ''.join(seqs)\n",
    "                name = line[1:]  # Skip '>'\n",
    "                seqs = []\n",
    "            else:\n",
    "                seqs.append(line)\n",
    "        if name:\n",
    "            yield name, ''.join(seqs)  # Yield the last entry\n",
    "\n",
    "# Function to read FASTQ file and yield name and sequence\n",
    "def fastq_iter(fastq_file):\n",
    "    with gzip.open(fastq_file, 'rt') as file:\n",
    "        while True:\n",
    "            name = file.readline().strip()[1:]  # Skip '@' in FASTQ\n",
    "            seq = file.readline().strip()\n",
    "            file.readline()  # Skip the '+' line\n",
    "            file.readline()  # Skip the quality line\n",
    "            if not name or not seq:\n",
    "                break\n",
    "            yield name, seq\n",
    "\n",
    "# Fungsi untuk menghitung frekuensi global k-mer\n",
    "def compute_global_kmer_abundance(seq_iter, kmer_len):\n",
    "    global_kmer_counts = Counter()\n",
    "    for h, seq in seq_iter:\n",
    "        norm_seq = str(seq).upper()\n",
    "        for i in range(len(norm_seq) - kmer_len + 1):\n",
    "            kmer = norm_seq[i:i+kmer_len]\n",
    "            if set(kmer) <= {'A', 'T', 'G', 'C'}:\n",
    "                global_kmer_counts[kmer] += 1\n",
    "    return global_kmer_counts\n",
    "\n",
    "# Function to generate feature mapping for k-mers with sequences as keys\n",
    "def generate_feature_mapping(kmer_len):\n",
    "    BASE_COMPLEMENT = {\"A\": \"T\", \"T\": \"A\", \"G\": \"C\", \"C\": \"G\"}\n",
    "    kmer_hash = {}\n",
    "    kmer_names = []\n",
    "    counter = 0\n",
    "    for kmer in product(\"ATGC\", repeat=kmer_len):\n",
    "        kmer = ''.join(kmer)\n",
    "        rev_compl = ''.join([BASE_COMPLEMENT[x] for x in reversed(kmer)])  # Generate reverse complement\n",
    "        if kmer not in kmer_hash and rev_compl not in kmer_hash:\n",
    "            # Use canonical k-mer (lexicographically smaller of kmer and rev_compl)\n",
    "            canonical_kmer = min(kmer, rev_compl)\n",
    "            kmer_hash[canonical_kmer] = counter\n",
    "            kmer_names.append(canonical_kmer)  # Store the k-mer for naming features\n",
    "            counter += 1\n",
    "    return kmer_hash, kmer_names, counter\n",
    "    \n",
    "# Function to detect file type based on file extension\n",
    "def detect_file_type(file_path):\n",
    "    ext = os.path.splitext(file_path)[-1].lower()  # Get file extension and convert to lowercase\n",
    "    # Handle gzipped files\n",
    "    if ext == '.gz':\n",
    "        ext = os.path.splitext(file_path[:-3])[-1].lower()  # Check the extension before \".gz\"\n",
    "        \n",
    "    if ext in ['.fasta', '.fa', '.fna']:\n",
    "        return 'fasta'\n",
    "    elif ext in ['.fastq', '.fq']:\n",
    "        return 'fastq'\n",
    "    else:\n",
    "        raise ValueError(\"Unsupported file extension. Please provide a valid FASTA or FASTQ file.\")\n",
    "        \n",
    "# Function to generate k-mer features from FASTA or FASTQ with abundance feature\n",
    "def generate_kmer_features(folder_path, length_threshold, kmer_len, split=False, split_threshold=0):\n",
    "    # List all files in the folder with .fna.gz extension\n",
    "    files = [os.path.join(folder_path, f) for f in os.listdir(folder_path) if f.endswith('.fna.gz')]\n",
    "    \n",
    "    kmer_dict, kmer_names, nr_features = generate_feature_mapping(kmer_len)\n",
    "    composition = OrderedDict()\n",
    "    abundance = OrderedDict()  # Dictionary to store abundance (total k-mer count per sequence)\n",
    "\n",
    "    for file_path in files:\n",
    "        # Automatically detect file type from file extension\n",
    "        file_type = detect_file_type(file_path)\n",
    "\n",
    "        # Select the appropriate file iterator based on the file type\n",
    "        if file_type == 'fasta':\n",
    "            seq_iter = fasta_iter(file_path)\n",
    "        elif file_type == 'fastq':\n",
    "            seq_iter = fastq_iter(file_path)\n",
    "        else:\n",
    "            raise ValueError(\"Unsupported file type. Use 'fasta' or 'fastq'.\")\n",
    "\n",
    "        def seq_list():\n",
    "            for h, seq in seq_iter:\n",
    "                # Ambil hanya ID sebelum karakter \"|\"\n",
    "                h = h.split('|')[0].strip()\n",
    "                \n",
    "                if not split:\n",
    "                    yield h, seq\n",
    "                elif len(seq) >= split_threshold:\n",
    "                    half = len(seq) // 2\n",
    "                    yield (h + '_1', seq[:half])\n",
    "                    yield (h + '_2', seq[half:])\n",
    "\n",
    "        for h, seq in seq_list():\n",
    "            if len(seq) < length_threshold:\n",
    "                continue\n",
    "            norm_seq = str(seq).upper()\n",
    "            kmers = [kmer_dict[norm_seq[i:i+kmer_len]]\n",
    "                     for i in range(len(norm_seq) - kmer_len + 1)\n",
    "                     if norm_seq[i:i+kmer_len] in kmer_dict]  # Ignore kmers with non-canonical bases\n",
    "            \n",
    "            # Count the abundance of k-mers and store in composition and abundance dictionaries\n",
    "            composition[f\"{file_path}_{h}\"] = np.bincount(np.array(kmers, dtype=np.int64), minlength=nr_features)\n",
    "            abundance[f\"{file_path}_{h}\"] = len(kmers)  # Total k-mer count for abundance\n",
    "\n",
    "    # Create DataFrame with k-mer names as column headers\n",
    "    df = pd.DataFrame.from_dict(composition, orient='index', dtype=float, columns=kmer_names)\n",
    "\n",
    "    # Add abundance column\n",
    "    df['abundance'] = pd.Series(abundance)\n",
    "\n",
    "    # Apply smoothing and normalize by row (sequence)\n",
    "    df.iloc[:, :-1] = df.iloc[:, :-1].apply(lambda x: x + 1e-5)  # Smoothing\n",
    "    df.iloc[:, :-1] = df.iloc[:, :-1].div(df.iloc[:, :-1].sum(axis=1), axis=0)  # Normalize by row (sequence), excluding abundance\n",
    "\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e8f6389-537c-4390-877a-8f960d1474ae",
   "metadata": {},
   "source": [
    "# Contigs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35ad17f4-ea82-4c31-b4ef-fad1fa8a64b8",
   "metadata": {},
   "source": [
    "## Feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "149aa058-c277-4dc2-9fe1-582b4a4aba59",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAAAA</th>\n",
       "      <th>AAAAT</th>\n",
       "      <th>AAAAG</th>\n",
       "      <th>AAAAC</th>\n",
       "      <th>AAATA</th>\n",
       "      <th>AAATT</th>\n",
       "      <th>AAATG</th>\n",
       "      <th>AAATC</th>\n",
       "      <th>AAAGA</th>\n",
       "      <th>AAAGT</th>\n",
       "      <th>...</th>\n",
       "      <th>CGGAG</th>\n",
       "      <th>CCTCG</th>\n",
       "      <th>CGACG</th>\n",
       "      <th>CCACG</th>\n",
       "      <th>CCCCG</th>\n",
       "      <th>CGCCG</th>\n",
       "      <th>CCGCG</th>\n",
       "      <th>CCAGG</th>\n",
       "      <th>CCCGG</th>\n",
       "      <th>abundance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1</th>\n",
       "      <td>4.716867e-08</td>\n",
       "      <td>4.716867e-08</td>\n",
       "      <td>4.716867e-08</td>\n",
       "      <td>4.716914e-03</td>\n",
       "      <td>4.716867e-08</td>\n",
       "      <td>4.716867e-08</td>\n",
       "      <td>4.716867e-08</td>\n",
       "      <td>4.716867e-08</td>\n",
       "      <td>4.716914e-03</td>\n",
       "      <td>4.716914e-03</td>\n",
       "      <td>...</td>\n",
       "      <td>9.433782e-03</td>\n",
       "      <td>4.716867e-08</td>\n",
       "      <td>4.716914e-03</td>\n",
       "      <td>4.716867e-08</td>\n",
       "      <td>4.716914e-03</td>\n",
       "      <td>0.014151</td>\n",
       "      <td>4.716867e-08</td>\n",
       "      <td>4.716867e-08</td>\n",
       "      <td>1.415065e-02</td>\n",
       "      <td>212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1</th>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936968e-03</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936968e-03</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936968e-03</td>\n",
       "      <td>3.936968e-03</td>\n",
       "      <td>0.031495</td>\n",
       "      <td>3.936968e-03</td>\n",
       "      <td>7.873896e-03</td>\n",
       "      <td>3.936968e-03</td>\n",
       "      <td>254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1</th>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936968e-03</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936968e-03</td>\n",
       "      <td>3.936968e-03</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>0.003937</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>7.873896e-03</td>\n",
       "      <td>3.936929e-08</td>\n",
       "      <td>254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1</th>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>2.281328e-02</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>7.604453e-03</td>\n",
       "      <td>0.015209</td>\n",
       "      <td>3.802245e-03</td>\n",
       "      <td>3.802207e-08</td>\n",
       "      <td>3.802245e-03</td>\n",
       "      <td>263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1</th>\n",
       "      <td>3.999918e-08</td>\n",
       "      <td>3.999918e-08</td>\n",
       "      <td>3.999918e-08</td>\n",
       "      <td>3.999918e-08</td>\n",
       "      <td>3.999918e-08</td>\n",
       "      <td>3.999918e-08</td>\n",
       "      <td>3.999918e-08</td>\n",
       "      <td>3.999918e-08</td>\n",
       "      <td>3.999958e-03</td>\n",
       "      <td>3.999918e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>3.999918e-08</td>\n",
       "      <td>7.999876e-03</td>\n",
       "      <td>3.999958e-03</td>\n",
       "      <td>3.999918e-08</td>\n",
       "      <td>7.999876e-03</td>\n",
       "      <td>0.012000</td>\n",
       "      <td>7.999876e-03</td>\n",
       "      <td>1.599971e-02</td>\n",
       "      <td>3.999958e-03</td>\n",
       "      <td>250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1</th>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>4.237239e-03</td>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>4.237239e-03</td>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>4.237239e-03</td>\n",
       "      <td>8.474435e-03</td>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>0.004237</td>\n",
       "      <td>4.237239e-03</td>\n",
       "      <td>4.237239e-03</td>\n",
       "      <td>4.237196e-08</td>\n",
       "      <td>236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1</th>\n",
       "      <td>4.048499e-08</td>\n",
       "      <td>4.048540e-03</td>\n",
       "      <td>4.048540e-03</td>\n",
       "      <td>4.048499e-08</td>\n",
       "      <td>4.048499e-08</td>\n",
       "      <td>4.048499e-08</td>\n",
       "      <td>4.048540e-03</td>\n",
       "      <td>4.048499e-08</td>\n",
       "      <td>4.048499e-08</td>\n",
       "      <td>4.048540e-03</td>\n",
       "      <td>...</td>\n",
       "      <td>4.048540e-03</td>\n",
       "      <td>4.048499e-08</td>\n",
       "      <td>4.048499e-08</td>\n",
       "      <td>4.048540e-03</td>\n",
       "      <td>4.048499e-08</td>\n",
       "      <td>0.028340</td>\n",
       "      <td>4.048540e-03</td>\n",
       "      <td>4.048499e-08</td>\n",
       "      <td>4.048499e-08</td>\n",
       "      <td>247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1</th>\n",
       "      <td>1.090892e-02</td>\n",
       "      <td>1.090892e-02</td>\n",
       "      <td>1.090892e-02</td>\n",
       "      <td>3.636296e-08</td>\n",
       "      <td>3.636296e-08</td>\n",
       "      <td>7.272628e-03</td>\n",
       "      <td>3.636332e-03</td>\n",
       "      <td>7.272628e-03</td>\n",
       "      <td>7.272628e-03</td>\n",
       "      <td>3.636296e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>3.636332e-03</td>\n",
       "      <td>3.636332e-03</td>\n",
       "      <td>3.636296e-08</td>\n",
       "      <td>3.636296e-08</td>\n",
       "      <td>3.636296e-08</td>\n",
       "      <td>0.003636</td>\n",
       "      <td>3.636332e-03</td>\n",
       "      <td>3.636332e-03</td>\n",
       "      <td>3.636296e-08</td>\n",
       "      <td>275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1</th>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>3.891012e-03</td>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>7.781985e-03</td>\n",
       "      <td>3.891012e-03</td>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>7.781985e-03</td>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>3.891012e-03</td>\n",
       "      <td>0.007782</td>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>3.890973e-08</td>\n",
       "      <td>3.891012e-03</td>\n",
       "      <td>257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1</th>\n",
       "      <td>3.689967e-08</td>\n",
       "      <td>3.690004e-03</td>\n",
       "      <td>3.689967e-08</td>\n",
       "      <td>3.690004e-03</td>\n",
       "      <td>1.106994e-02</td>\n",
       "      <td>3.689967e-08</td>\n",
       "      <td>3.689967e-08</td>\n",
       "      <td>3.690004e-03</td>\n",
       "      <td>3.690004e-03</td>\n",
       "      <td>3.689967e-08</td>\n",
       "      <td>...</td>\n",
       "      <td>3.689967e-08</td>\n",
       "      <td>3.690004e-03</td>\n",
       "      <td>1.106994e-02</td>\n",
       "      <td>3.689967e-08</td>\n",
       "      <td>3.690004e-03</td>\n",
       "      <td>0.011070</td>\n",
       "      <td>3.690004e-03</td>\n",
       "      <td>3.690004e-03</td>\n",
       "      <td>7.379971e-03</td>\n",
       "      <td>271</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100000 rows × 513 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                       AAAAA         AAAAT  \\\n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1      4.716867e-08  4.716867e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1      3.936929e-08  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1      3.936929e-08  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1      3.802207e-08  3.802207e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1      3.999918e-08  3.999918e-08   \n",
       "...                                                      ...           ...   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1   4.237196e-08  4.237196e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1   4.048499e-08  4.048540e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1   1.090892e-02  1.090892e-02   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1   3.890973e-08  3.891012e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1  3.689967e-08  3.690004e-03   \n",
       "\n",
       "                                                       AAAAG         AAAAC  \\\n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1      4.716867e-08  4.716914e-03   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1      3.936968e-03  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1      3.936929e-08  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1      3.802207e-08  3.802207e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1      3.999918e-08  3.999918e-08   \n",
       "...                                                      ...           ...   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1   4.237196e-08  4.237239e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1   4.048540e-03  4.048499e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1   1.090892e-02  3.636296e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1   3.890973e-08  3.890973e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1  3.689967e-08  3.690004e-03   \n",
       "\n",
       "                                                       AAATA         AAATT  \\\n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1      4.716867e-08  4.716867e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1      3.936929e-08  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1      3.936929e-08  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1      3.802207e-08  3.802207e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1      3.999918e-08  3.999918e-08   \n",
       "...                                                      ...           ...   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1   4.237196e-08  4.237239e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1   4.048499e-08  4.048499e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1   3.636296e-08  7.272628e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1   7.781985e-03  3.891012e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1  1.106994e-02  3.689967e-08   \n",
       "\n",
       "                                                       AAATG         AAATC  \\\n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1      4.716867e-08  4.716867e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1      3.936929e-08  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1      3.936929e-08  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1      3.802207e-08  3.802207e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1      3.999918e-08  3.999918e-08   \n",
       "...                                                      ...           ...   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1   4.237196e-08  4.237239e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1   4.048540e-03  4.048499e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1   3.636332e-03  7.272628e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1   3.890973e-08  3.890973e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1  3.689967e-08  3.690004e-03   \n",
       "\n",
       "                                                       AAAGA         AAAGT  \\\n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1      4.716914e-03  4.716914e-03   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1      3.936968e-03  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1      3.936968e-03  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1      3.802207e-08  3.802207e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1      3.999958e-03  3.999918e-08   \n",
       "...                                                      ...           ...   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1   8.474435e-03  4.237196e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1   4.048499e-08  4.048540e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1   7.272628e-03  3.636296e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1   3.890973e-08  3.890973e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1  3.690004e-03  3.689967e-08   \n",
       "\n",
       "                                                ...         CGGAG  \\\n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1      ...  9.433782e-03   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1      ...  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1      ...  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1      ...  3.802207e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1      ...  3.999918e-08   \n",
       "...                                             ...           ...   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1   ...  4.237196e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1   ...  4.048540e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1   ...  3.636332e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1   ...  7.781985e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1  ...  3.689967e-08   \n",
       "\n",
       "                                                       CCTCG         CGACG  \\\n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1      4.716867e-08  4.716914e-03   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1      3.936929e-08  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1      3.936968e-03  3.936968e-03   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1      3.802207e-08  2.281328e-02   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1      7.999876e-03  3.999958e-03   \n",
       "...                                                      ...           ...   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1   4.237196e-08  4.237196e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1   4.048499e-08  4.048499e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1   3.636332e-03  3.636296e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1   3.890973e-08  3.890973e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1  3.690004e-03  1.106994e-02   \n",
       "\n",
       "                                                       CCACG         CCCCG  \\\n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1      4.716867e-08  4.716914e-03   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1      3.936968e-03  3.936968e-03   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1      3.936929e-08  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1      3.802207e-08  7.604453e-03   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1      3.999918e-08  7.999876e-03   \n",
       "...                                                      ...           ...   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1   4.237196e-08  4.237196e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1   4.048540e-03  4.048499e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1   3.636296e-08  3.636296e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1   3.890973e-08  3.891012e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1  3.689967e-08  3.690004e-03   \n",
       "\n",
       "                                                   CGCCG         CCGCG  \\\n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1      0.014151  4.716867e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1      0.031495  3.936968e-03   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1      0.003937  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1      0.015209  3.802245e-03   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1      0.012000  7.999876e-03   \n",
       "...                                                  ...           ...   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1   0.004237  4.237239e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1   0.028340  4.048540e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1   0.003636  3.636332e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1   0.007782  3.890973e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1  0.011070  3.690004e-03   \n",
       "\n",
       "                                                       CCAGG         CCCGG  \\\n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1      4.716867e-08  1.415065e-02   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1      7.873896e-03  3.936968e-03   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1      7.873896e-03  3.936929e-08   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1      3.802207e-08  3.802245e-03   \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1      1.599971e-02  3.999958e-03   \n",
       "...                                                      ...           ...   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1   4.237239e-03  4.237196e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1   4.048499e-08  4.048499e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1   3.636332e-03  3.636296e-08   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1   3.890973e-08  3.891012e-03   \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1  3.690004e-03  7.379971e-03   \n",
       "\n",
       "                                                abundance  \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r1.1            212  \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r2.1            254  \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r3.1            254  \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r4.1            263  \n",
       "./GCF_000009905-Exact.d324fd4b.fna.gz_r5.1            250  \n",
       "...                                                   ...  \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9996.1         236  \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9997.1         247  \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9998.1         275  \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r9999.1         257  \n",
       "./GCF_000012985-Exact.d32508a2.fna.gz_r10000.1        271  \n",
       "\n",
       "[100000 rows x 513 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_folder = './'  # Updated file path to your uploaded file\n",
    "kmer_len = 5  # Original k-mer length\n",
    "length_threshold = 100  # Example sequence length threshold\n",
    "\n",
    "# Generate spaced k-mer features\n",
    "kmer_df = generate_kmer_features(fasta_folder, length_threshold, kmer_len)\n",
    "kmer_df = pd.DataFrame(kmer_df)\n",
    "kmer_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b030146f-3502-49d6-ad5a-1e5170395687",
   "metadata": {},
   "source": [
    "## Label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ccc77d20-9b15-4645-9fb7-887c1bfa9ebc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_006177.1 Symbiobacterium thermophilum IAM 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_006177.1 Symbiobacterium thermophilum IAM 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_006177.1 Symbiobacterium thermophilum IAM 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_006177.1 Symbiobacterium thermophilum IAM 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_006177.1 Symbiobacterium thermophilum IAM 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99995</th>\n",
       "      <td>NC_007645.1 Hahella chejuensis KCTC 2396, comp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99996</th>\n",
       "      <td>NC_007645.1 Hahella chejuensis KCTC 2396, comp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99997</th>\n",
       "      <td>NC_007645.1 Hahella chejuensis KCTC 2396, comp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99998</th>\n",
       "      <td>NC_007645.1 Hahella chejuensis KCTC 2396, comp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99999</th>\n",
       "      <td>NC_007645.1 Hahella chejuensis KCTC 2396, comp...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100000 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   Label\n",
       "0      NC_006177.1 Symbiobacterium thermophilum IAM 1...\n",
       "1      NC_006177.1 Symbiobacterium thermophilum IAM 1...\n",
       "2      NC_006177.1 Symbiobacterium thermophilum IAM 1...\n",
       "3      NC_006177.1 Symbiobacterium thermophilum IAM 1...\n",
       "4      NC_006177.1 Symbiobacterium thermophilum IAM 1...\n",
       "...                                                  ...\n",
       "99995  NC_007645.1 Hahella chejuensis KCTC 2396, comp...\n",
       "99996  NC_007645.1 Hahella chejuensis KCTC 2396, comp...\n",
       "99997  NC_007645.1 Hahella chejuensis KCTC 2396, comp...\n",
       "99998  NC_007645.1 Hahella chejuensis KCTC 2396, comp...\n",
       "99999  NC_007645.1 Hahella chejuensis KCTC 2396, comp...\n",
       "\n",
       "[100000 rows x 1 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# label_path = \"./contigs/gsa_mapping.tsv.gz\"\n",
    "# label_df = pd.read_csv(label_path, sep='\\t')\n",
    "# label_df = label_df[[\"genome_id\"]]\n",
    "# label_df\n",
    "\n",
    "def generate_label_dataframe(file_path):\n",
    "    labels = OrderedDict()\n",
    "    \n",
    "    # Automatically detect file type\n",
    "    file_type = detect_file_type(file_path)\n",
    "    if file_type == 'fasta':\n",
    "        seq_iter = fasta_iter(file_path)\n",
    "    elif file_type == 'fastq':\n",
    "        seq_iter = fastq_iter(file_path)\n",
    "    else:\n",
    "        raise ValueError(\"Unsupported file type. Use 'fasta' or 'fastq'.\")\n",
    "\n",
    "    for h, seq in seq_iter:\n",
    "        # Ambil hanya ID sebelum karakter \"|\"\n",
    "        seq_id = h.split('|')[0].strip()\n",
    "        # Ekstrak label di antara tanda petik dua (\"\")\n",
    "        label_match = re.search(r'\"(.*?)\"', h)\n",
    "        label = label_match.group(1) if label_match else \"Unknown\"\n",
    "        \n",
    "        # Simpan ID dan label\n",
    "        labels[seq_id] = label\n",
    "\n",
    "    # Buat DataFrame dari dictionary labels\n",
    "    label_df = pd.DataFrame.from_dict(labels, orient='index', columns=['Label'])\n",
    "    \n",
    "    return label_df\n",
    "\n",
    "folder_path = './'\n",
    "files = [os.path.join(folder_path, f) for f in os.listdir(folder_path) if f.endswith('.fna.gz')]\n",
    "\n",
    "# Menggabungkan semua label dari beberapa file\n",
    "label_df = pd.concat([generate_label_dataframe(f) for f in files])\n",
    "\n",
    "# Reset indeks jika diperlukan\n",
    "label_df.reset_index(inplace=True)\n",
    "label_df.columns = ['Sequence_ID', 'Label']\n",
    "label_df =  label_df[['Label']]\n",
    "label_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2e12dff7-c5fe-4b9c-8a3e-7240c214b8d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Statistik Deskriptif Data Fitur:\n",
      "              AAAAA         AAAAT         AAAAG         AAAAC         AAATA  \\\n",
      "count  1.000000e+05  1.000000e+05  1.000000e+05  1.000000e+05  1.000000e+05   \n",
      "mean   7.192815e-03  5.690635e-03  4.235023e-03  3.654365e-03  4.729090e-03   \n",
      "std    8.771448e-03  6.020118e-03  4.864990e-03  4.166099e-03  5.601509e-03   \n",
      "min    2.481358e-08  2.358462e-08  2.358462e-08  2.577286e-08  2.358462e-08   \n",
      "25%    4.328908e-08  4.405187e-08  4.255226e-08  4.219318e-08  4.201590e-08   \n",
      "50%    4.149331e-03  4.166619e-03  3.787843e-03  3.662972e-03  3.846117e-03   \n",
      "75%    1.153827e-02  8.695502e-03  7.352840e-03  5.434686e-03  7.812383e-03   \n",
      "max    1.051199e-01  4.498194e-02  4.297934e-02  3.401305e-02  5.395588e-02   \n",
      "\n",
      "              AAATT         AAATG         AAATC         AAAGA         AAAGT  \\\n",
      "count  1.000000e+05  1.000000e+05  1.000000e+05  1.000000e+05  1.000000e+05   \n",
      "mean   4.254107e-03  3.268449e-03  3.496504e-03  3.605722e-03  2.634730e-03   \n",
      "std    5.126718e-03  4.091042e-03  4.054192e-03  4.552977e-03  3.570478e-03   \n",
      "min    2.358462e-08  2.358462e-08  2.358462e-08  2.358462e-08  2.358462e-08   \n",
      "25%    4.149289e-08  4.098275e-08  4.149289e-08  4.166578e-08  4.015982e-08   \n",
      "50%    3.717439e-03  3.257307e-03  3.597092e-03  3.436400e-03  4.784572e-08   \n",
      "75%    7.352840e-03  4.587094e-03  4.807622e-03  4.999922e-03  4.219360e-03   \n",
      "max    4.310254e-02  3.383397e-02  4.532515e-02  5.444049e-02  3.076866e-02   \n",
      "\n",
      "       ...         CGGAG         CCTCG         CGACG         CCACG  \\\n",
      "count  ...  1.000000e+05  1.000000e+05  1.000000e+05  1.000000e+05   \n",
      "mean   ...  1.476732e-03  1.260728e-03  1.468669e-03  1.297330e-03   \n",
      "std    ...  2.879007e-03  2.806659e-03  3.062545e-03  2.706591e-03   \n",
      "min    ...  2.481358e-08  2.358462e-08  2.577286e-08  2.358462e-08   \n",
      "25%    ...  3.860928e-08  3.831342e-08  3.860928e-08  3.846078e-08   \n",
      "50%    ...  4.273411e-08  4.201590e-08  4.255226e-08  4.237196e-08   \n",
      "75%    ...  3.496476e-03  5.181210e-08  6.493291e-08  5.555398e-08   \n",
      "max    ...  6.856948e-02  6.382866e-02  8.558366e-02  3.964673e-02   \n",
      "\n",
      "              CCCCG         CGCCG         CCGCG         CCAGG         CCCGG  \\\n",
      "count  1.000000e+05  1.000000e+05  1.000000e+05  1.000000e+05  1.000000e+05   \n",
      "mean   1.542716e-03  3.374769e-03  1.743653e-03  2.283564e-03  2.462758e-03   \n",
      "std    3.127012e-03  5.871782e-03  3.477454e-03  3.966292e-03  4.561096e-03   \n",
      "min    2.358462e-08  2.577286e-08  2.358462e-08  2.577286e-08  2.358462e-08   \n",
      "25%    3.860928e-08  3.936929e-08  3.860928e-08  3.921490e-08  3.890973e-08   \n",
      "50%    4.273411e-08  4.504401e-08  4.291751e-08  4.444343e-08  4.405187e-08   \n",
      "75%    3.436400e-03  4.310293e-03  3.623157e-03  3.968213e-03  3.984022e-03   \n",
      "max    5.485118e-02  7.692097e-02  3.718934e-02  3.813481e-02  1.101648e-01   \n",
      "\n",
      "           abundance  \n",
      "count  100000.000000  \n",
      "mean      247.926130  \n",
      "std        29.084667  \n",
      "min        79.000000  \n",
      "25%       229.000000  \n",
      "50%       248.000000  \n",
      "75%       267.000000  \n",
      "max       424.000000  \n",
      "\n",
      "[8 rows x 513 columns]\n",
      "\n",
      "Distribusi Kelas pada Data Label:\n",
      "Label                                                                                  \n",
      "NC_000964.3 Bacillus subtilis subsp. subtilis str. 168 complete genome                     10000\n",
      "NC_003910.7 Colwellia psychrerythraea 34H, complete sequence                               10000\n",
      "NC_006177.1 Symbiobacterium thermophilum IAM 14863, complete sequence                      10000\n",
      "NC_007645.1 Hahella chejuensis KCTC 2396, complete sequence                                10000\n",
      "NC_007795.1 Staphylococcus aureus subsp. aureus NCTC 8325 chromosome, complete genome      10000\n",
      "NC_008346.1 Syntrophomonas wolfei subsp. wolfei str. Goettingen G311, complete sequence    10000\n",
      "NC_008709.1 Psychromonas ingrahamii 37, complete sequence                                  10000\n",
      "NC_009922.1 Alkaliphilus oremlandii OhILAs, complete sequence                              10000\n",
      "NC_011830.1 Desulfitobacterium hafniense DCB-2, complete sequence                          10000\n",
      "NC_013422.1 Halothiobacillus neapolitanus c2, complete sequence                            10000\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Tampilkan statistik deskriptif dari data fitur\n",
    "print(\"Statistik Deskriptif Data Fitur:\")\n",
    "print(kmer_df.describe())\n",
    "\n",
    "# Tampilkan distribusi kelas dari data label\n",
    "print(\"\\nDistribusi Kelas pada Data Label:\")\n",
    "print(label_df.value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "062a483a-03d4-490d-9f85-c500982cf6f1",
   "metadata": {},
   "source": [
    "## Classifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d96efc46-054f-46eb-80e1-d7093195c0be",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split\n",
    "import time\n",
    "from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.neighbors import KNeighborsClassifier\n",
    "from sklearn.neural_network import MLPClassifier\n",
    "from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, classification_report"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7517d6b0-cbd9-494b-ab9a-3941c5270bf1",
   "metadata": {},
   "source": [
    "### Pembagian Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "2f17964d-9334-4e9f-b9dc-c8fc1a264e3c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Jumlah data latih per kelas:\n",
      "(array(['NC_000964.3 Bacillus subtilis subsp. subtilis str. 168 complete genome',\n",
      "       'NC_003910.7 Colwellia psychrerythraea 34H, complete sequence',\n",
      "       'NC_006177.1 Symbiobacterium thermophilum IAM 14863, complete sequence',\n",
      "       'NC_007645.1 Hahella chejuensis KCTC 2396, complete sequence',\n",
      "       'NC_007795.1 Staphylococcus aureus subsp. aureus NCTC 8325 chromosome, complete genome',\n",
      "       'NC_008346.1 Syntrophomonas wolfei subsp. wolfei str. Goettingen G311, complete sequence',\n",
      "       'NC_008709.1 Psychromonas ingrahamii 37, complete sequence',\n",
      "       'NC_009922.1 Alkaliphilus oremlandii OhILAs, complete sequence',\n",
      "       'NC_011830.1 Desulfitobacterium hafniense DCB-2, complete sequence',\n",
      "       'NC_013422.1 Halothiobacillus neapolitanus c2, complete sequence'],\n",
      "      dtype=object), array([8023, 7962, 7995, 8014, 8008, 8000, 7998, 7987, 8015, 7998]))\n",
      "\n",
      "Jumlah data uji per kelas:\n",
      "(array(['NC_000964.3 Bacillus subtilis subsp. subtilis str. 168 complete genome',\n",
      "       'NC_003910.7 Colwellia psychrerythraea 34H, complete sequence',\n",
      "       'NC_006177.1 Symbiobacterium thermophilum IAM 14863, complete sequence',\n",
      "       'NC_007645.1 Hahella chejuensis KCTC 2396, complete sequence',\n",
      "       'NC_007795.1 Staphylococcus aureus subsp. aureus NCTC 8325 chromosome, complete genome',\n",
      "       'NC_008346.1 Syntrophomonas wolfei subsp. wolfei str. Goettingen G311, complete sequence',\n",
      "       'NC_008709.1 Psychromonas ingrahamii 37, complete sequence',\n",
      "       'NC_009922.1 Alkaliphilus oremlandii OhILAs, complete sequence',\n",
      "       'NC_011830.1 Desulfitobacterium hafniense DCB-2, complete sequence',\n",
      "       'NC_013422.1 Halothiobacillus neapolitanus c2, complete sequence'],\n",
      "      dtype=object), array([1977, 2038, 2005, 1986, 1992, 2000, 2002, 2013, 1985, 2002]))\n"
     ]
    }
   ],
   "source": [
    "features = kmer_df\n",
    "labels = label_df\n",
    "\n",
    "# Pisahkan data menjadi training dan testing set (80% training, 20% testing)\n",
    "X_train, X_test, y_train, y_test = train_test_split(features, labels, test_size=0.2, random_state=42)\n",
    "\n",
    "print(\"Jumlah data latih per kelas:\")\n",
    "print(np.unique(y_train, return_counts=True))\n",
    "\n",
    "print(\"\\nJumlah data uji per kelas:\")\n",
    "print(np.unique(y_test, return_counts=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e81fefcd-0a21-4c92-9d9b-0716da49763f",
   "metadata": {},
   "source": [
    "### Random Forest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e9b159a5-3c65-4175-b48b-c26bf0cf0712",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/sklearn/base.py:1473: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().\n",
      "  return fit_method(estimator, *args, **kwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Classification Report untuk Random Forest:\n",
      "                                                                                         precision    recall  f1-score   support\n",
      "\n",
      "                 NC_000964.3 Bacillus subtilis subsp. subtilis str. 168 complete genome       0.74      0.71      0.73      1977\n",
      "                           NC_003910.7 Colwellia psychrerythraea 34H, complete sequence       0.66      0.74      0.70      2038\n",
      "                  NC_006177.1 Symbiobacterium thermophilum IAM 14863, complete sequence       0.98      0.96      0.97      2005\n",
      "                            NC_007645.1 Hahella chejuensis KCTC 2396, complete sequence       0.81      0.78      0.80      1986\n",
      "  NC_007795.1 Staphylococcus aureus subsp. aureus NCTC 8325 chromosome, complete genome       0.74      0.88      0.80      1992\n",
      "NC_008346.1 Syntrophomonas wolfei subsp. wolfei str. Goettingen G311, complete sequence       0.75      0.77      0.76      2000\n",
      "                              NC_008709.1 Psychromonas ingrahamii 37, complete sequence       0.69      0.53      0.60      2002\n",
      "                          NC_009922.1 Alkaliphilus oremlandii OhILAs, complete sequence       0.76      0.84      0.80      2013\n",
      "                      NC_011830.1 Desulfitobacterium hafniense DCB-2, complete sequence       0.74      0.64      0.69      1985\n",
      "                        NC_013422.1 Halothiobacillus neapolitanus c2, complete sequence       0.86      0.87      0.87      2002\n",
      "\n",
      "                                                                               accuracy                           0.77     20000\n",
      "                                                                              macro avg       0.77      0.77      0.77     20000\n",
      "                                                                           weighted avg       0.77      0.77      0.77     20000\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/sklearn/utils/validation.py:1339: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().\n",
      "  y = column_or_1d(y, warn=True)\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):\n",
      "STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.\n",
      "\n",
      "Increase the number of iterations (max_iter) or scale the data as shown in:\n",
      "    https://scikit-learn.org/stable/modules/preprocessing.html\n",
      "Please also refer to the documentation for alternative solver options:\n",
      "    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression\n",
      "  n_iter_i = _check_optimize_result(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Classification Report untuk Logistic Regression:\n",
      "                                                                                         precision    recall  f1-score   support\n",
      "\n",
      "                 NC_000964.3 Bacillus subtilis subsp. subtilis str. 168 complete genome       0.69      0.63      0.65      1977\n",
      "                           NC_003910.7 Colwellia psychrerythraea 34H, complete sequence       0.57      0.55      0.56      2038\n",
      "                  NC_006177.1 Symbiobacterium thermophilum IAM 14863, complete sequence       0.91      0.95      0.93      2005\n",
      "                            NC_007645.1 Hahella chejuensis KCTC 2396, complete sequence       0.77      0.71      0.74      1986\n",
      "  NC_007795.1 Staphylococcus aureus subsp. aureus NCTC 8325 chromosome, complete genome       0.52      0.80      0.63      1992\n",
      "NC_008346.1 Syntrophomonas wolfei subsp. wolfei str. Goettingen G311, complete sequence       0.57      0.55      0.56      2000\n",
      "                              NC_008709.1 Psychromonas ingrahamii 37, complete sequence       0.54      0.37      0.44      2002\n",
      "                          NC_009922.1 Alkaliphilus oremlandii OhILAs, complete sequence       0.64      0.63      0.63      2013\n",
      "                      NC_011830.1 Desulfitobacterium hafniense DCB-2, complete sequence       0.62      0.51      0.56      1985\n",
      "                        NC_013422.1 Halothiobacillus neapolitanus c2, complete sequence       0.73      0.83      0.78      2002\n",
      "\n",
      "                                                                               accuracy                           0.65     20000\n",
      "                                                                              macro avg       0.65      0.65      0.65     20000\n",
      "                                                                           weighted avg       0.65      0.65      0.65     20000\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/sklearn/utils/validation.py:1339: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().\n",
      "  y = column_or_1d(y, warn=True)\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Classification Report untuk Support Vector Machine:\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                                                         precision    recall  f1-score   support\n",
      "\n",
      "                 NC_000964.3 Bacillus subtilis subsp. subtilis str. 168 complete genome       0.13      0.06      0.08      1977\n",
      "                           NC_003910.7 Colwellia psychrerythraea 34H, complete sequence       0.00      0.00      0.00      2038\n",
      "                  NC_006177.1 Symbiobacterium thermophilum IAM 14863, complete sequence       0.00      0.00      0.00      2005\n",
      "                            NC_007645.1 Hahella chejuensis KCTC 2396, complete sequence       0.10      0.03      0.04      1986\n",
      "  NC_007795.1 Staphylococcus aureus subsp. aureus NCTC 8325 chromosome, complete genome       0.16      0.04      0.06      1992\n",
      "NC_008346.1 Syntrophomonas wolfei subsp. wolfei str. Goettingen G311, complete sequence       0.10      0.90      0.19      2000\n",
      "                              NC_008709.1 Psychromonas ingrahamii 37, complete sequence       0.00      0.00      0.00      2002\n",
      "                          NC_009922.1 Alkaliphilus oremlandii OhILAs, complete sequence       0.15      0.02      0.03      2013\n",
      "                      NC_011830.1 Desulfitobacterium hafniense DCB-2, complete sequence       0.05      0.01      0.01      1985\n",
      "                        NC_013422.1 Halothiobacillus neapolitanus c2, complete sequence       0.14      0.01      0.02      2002\n",
      "\n",
      "                                                                               accuracy                           0.11     20000\n",
      "                                                                              macro avg       0.08      0.11      0.04     20000\n",
      "                                                                           weighted avg       0.08      0.11      0.04     20000\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/sklearn/neighbors/_classification.py:238: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().\n",
      "  return self._fit(X, y)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Classification Report untuk K-Nearest Neighbors:\n",
      "                                                                                         precision    recall  f1-score   support\n",
      "\n",
      "                 NC_000964.3 Bacillus subtilis subsp. subtilis str. 168 complete genome       0.43      0.61      0.50      1977\n",
      "                           NC_003910.7 Colwellia psychrerythraea 34H, complete sequence       0.42      0.57      0.48      2038\n",
      "                  NC_006177.1 Symbiobacterium thermophilum IAM 14863, complete sequence       0.94      0.92      0.93      2005\n",
      "                            NC_007645.1 Hahella chejuensis KCTC 2396, complete sequence       0.61      0.61      0.61      1986\n",
      "  NC_007795.1 Staphylococcus aureus subsp. aureus NCTC 8325 chromosome, complete genome       0.57      0.61      0.59      1992\n",
      "NC_008346.1 Syntrophomonas wolfei subsp. wolfei str. Goettingen G311, complete sequence       0.53      0.47      0.50      2000\n",
      "                              NC_008709.1 Psychromonas ingrahamii 37, complete sequence       0.39      0.32      0.35      2002\n",
      "                          NC_009922.1 Alkaliphilus oremlandii OhILAs, complete sequence       0.57      0.47      0.52      2013\n",
      "                      NC_011830.1 Desulfitobacterium hafniense DCB-2, complete sequence       0.51      0.35      0.41      1985\n",
      "                        NC_013422.1 Halothiobacillus neapolitanus c2, complete sequence       0.72      0.67      0.70      2002\n",
      "\n",
      "                                                                               accuracy                           0.56     20000\n",
      "                                                                              macro avg       0.57      0.56      0.56     20000\n",
      "                                                                           weighted avg       0.57      0.56      0.56     20000\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model</th>\n",
       "      <th>Accuracy</th>\n",
       "      <th>Precision</th>\n",
       "      <th>Recall</th>\n",
       "      <th>F1 Score</th>\n",
       "      <th>Time (seconds)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Random Forest</td>\n",
       "      <td>0.77280</td>\n",
       "      <td>0.772993</td>\n",
       "      <td>0.77280</td>\n",
       "      <td>0.770303</td>\n",
       "      <td>139.039397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Logistic Regression</td>\n",
       "      <td>0.65310</td>\n",
       "      <td>0.654736</td>\n",
       "      <td>0.65310</td>\n",
       "      <td>0.648167</td>\n",
       "      <td>144.464875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Support Vector Machine</td>\n",
       "      <td>0.10615</td>\n",
       "      <td>0.082078</td>\n",
       "      <td>0.10615</td>\n",
       "      <td>0.043710</td>\n",
       "      <td>4011.303041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>K-Nearest Neighbors</td>\n",
       "      <td>0.56100</td>\n",
       "      <td>0.568342</td>\n",
       "      <td>0.56100</td>\n",
       "      <td>0.559343</td>\n",
       "      <td>48.983661</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Model  Accuracy  Precision   Recall  F1 Score  \\\n",
       "0           Random Forest   0.77280   0.772993  0.77280  0.770303   \n",
       "1     Logistic Regression   0.65310   0.654736  0.65310  0.648167   \n",
       "2  Support Vector Machine   0.10615   0.082078  0.10615  0.043710   \n",
       "3     K-Nearest Neighbors   0.56100   0.568342  0.56100  0.559343   \n",
       "\n",
       "   Time (seconds)  \n",
       "0      139.039397  \n",
       "1      144.464875  \n",
       "2     4011.303041  \n",
       "3       48.983661  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# List model yang akan digunakan\n",
    "models = {\n",
    "    \"Random Forest\": RandomForestClassifier(n_estimators=100, random_state=42),\n",
    "    \"Logistic Regression\": LogisticRegression(random_state=42),\n",
    "    \"Support Vector Machine\": SVC(random_state=42),\n",
    "    \"K-Nearest Neighbors\": KNeighborsClassifier(),\n",
    "    #\"Gradient Boosting\": GradientBoostingClassifier(n_estimators=100, random_state=42),\n",
    "    #\"MLP Classifier\": MLPClassifier(random_state=42, max_iter=500)\n",
    "}\n",
    "\n",
    "# Menyimpan hasil metrik dan waktu komputasi\n",
    "results = []\n",
    "\n",
    "# Loop melalui semua model dan evaluasi performanya\n",
    "for model_name, model in models.items():\n",
    "    start_time = time.time()  # Mulai penghitungan waktu\n",
    "    \n",
    "    # Latih model\n",
    "    model.fit(X_train, y_train)\n",
    "    \n",
    "    # Prediksi menggunakan data testing\n",
    "    y_pred = model.predict(X_test)\n",
    "    \n",
    "    # Hitung waktu komputasi\n",
    "    elapsed_time = time.time() - start_time\n",
    "    \n",
    "    # Evaluasi model dengan berbagai metrik\n",
    "    accuracy = accuracy_score(y_test, y_pred)\n",
    "    precision = precision_score(y_test, y_pred, average='weighted')\n",
    "    recall = recall_score(y_test, y_pred, average='weighted')\n",
    "    f1 = f1_score(y_test, y_pred, average='weighted')\n",
    "    \n",
    "    # Simpan hasil dalam dictionary\n",
    "    results.append({\n",
    "        \"Model\": model_name,\n",
    "        \"Accuracy\": accuracy,\n",
    "        \"Precision\": precision,\n",
    "        \"Recall\": recall,\n",
    "        \"F1 Score\": f1,\n",
    "        \"Time (seconds)\": elapsed_time\n",
    "    })\n",
    "    \n",
    "    # Tampilkan hasil classification report untuk setiap model\n",
    "    print(f\"\\nClassification Report untuk {model_name}:\")\n",
    "    print(classification_report(y_test, y_pred))\n",
    "\n",
    "# Tampilkan hasil dalam bentuk dataframe untuk mempermudah pembacaan\n",
    "import pandas as pd\n",
    "results_df = pd.DataFrame(results)\n",
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "93f32311-1b80-45e2-a60e-96a7bc7488ee",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:1105: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().\n",
      "  y = column_or_1d(y, warn=True)\n",
      "/opt/conda/lib/python3.11/site-packages/sklearn/neural_network/_multilayer_perceptron.py:690: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (100) reached and the optimization hasn't converged yet.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Classification Report untuk MLP Classifier:\n",
      "                                                                                         precision    recall  f1-score   support\n",
      "\n",
      "                 NC_000964.3 Bacillus subtilis subsp. subtilis str. 168 complete genome       0.82      0.83      0.82      1977\n",
      "                           NC_003910.7 Colwellia psychrerythraea 34H, complete sequence       0.85      0.67      0.75      2038\n",
      "                  NC_006177.1 Symbiobacterium thermophilum IAM 14863, complete sequence       1.00      0.94      0.97      2005\n",
      "                            NC_007645.1 Hahella chejuensis KCTC 2396, complete sequence       0.85      0.89      0.87      1986\n",
      "  NC_007795.1 Staphylococcus aureus subsp. aureus NCTC 8325 chromosome, complete genome       0.79      0.92      0.85      1992\n",
      "NC_008346.1 Syntrophomonas wolfei subsp. wolfei str. Goettingen G311, complete sequence       0.77      0.84      0.80      2000\n",
      "                              NC_008709.1 Psychromonas ingrahamii 37, complete sequence       0.66      0.76      0.71      2002\n",
      "                          NC_009922.1 Alkaliphilus oremlandii OhILAs, complete sequence       0.87      0.85      0.86      2013\n",
      "                      NC_011830.1 Desulfitobacterium hafniense DCB-2, complete sequence       0.82      0.72      0.76      1985\n",
      "                        NC_013422.1 Halothiobacillus neapolitanus c2, complete sequence       0.93      0.89      0.91      2002\n",
      "\n",
      "                                                                               accuracy                           0.83     20000\n",
      "                                                                              macro avg       0.84      0.83      0.83     20000\n",
      "                                                                           weighted avg       0.84      0.83      0.83     20000\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model</th>\n",
       "      <th>Accuracy</th>\n",
       "      <th>Precision</th>\n",
       "      <th>Recall</th>\n",
       "      <th>F1 Score</th>\n",
       "      <th>Time (seconds)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MLP Classifier</td>\n",
       "      <td>0.83</td>\n",
       "      <td>0.835395</td>\n",
       "      <td>0.83</td>\n",
       "      <td>0.830154</td>\n",
       "      <td>615.483605</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Model  Accuracy  Precision  Recall  F1 Score  Time (seconds)\n",
       "0  MLP Classifier      0.83   0.835395    0.83  0.830154      615.483605"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# List model yang akan digunakan\n",
    "models = {\n",
    "    # \"Random Forest\": RandomForestClassifier(n_estimators=100, random_state=42),\n",
    "    # \"Logistic Regression\": LogisticRegression(random_state=42),\n",
    "    # \"Support Vector Machine\": SVC(random_state=42),\n",
    "    # \"K-Nearest Neighbors\": KNeighborsClassifier(),\n",
    "    # \"Gradient Boosting\": GradientBoostingClassifier(n_estimators=100, random_state=42),\n",
    "    \"MLP Classifier\": MLPClassifier(random_state=42, max_iter=100)\n",
    "}\n",
    "\n",
    "# Menyimpan hasil metrik dan waktu komputasi\n",
    "results = []\n",
    "\n",
    "# Loop melalui semua model dan evaluasi performanya\n",
    "for model_name, model in models.items():\n",
    "    start_time = time.time()  # Mulai penghitungan waktu\n",
    "    \n",
    "    # Latih model\n",
    "    model.fit(X_train, y_train)\n",
    "    \n",
    "    # Prediksi menggunakan data testing\n",
    "    y_pred = model.predict(X_test)\n",
    "    \n",
    "    # Hitung waktu komputasi\n",
    "    elapsed_time = time.time() - start_time\n",
    "    \n",
    "    # Evaluasi model dengan berbagai metrik\n",
    "    accuracy = accuracy_score(y_test, y_pred)\n",
    "    precision = precision_score(y_test, y_pred, average='weighted')\n",
    "    recall = recall_score(y_test, y_pred, average='weighted')\n",
    "    f1 = f1_score(y_test, y_pred, average='weighted')\n",
    "    \n",
    "    # Simpan hasil dalam dictionary\n",
    "    results.append({\n",
    "        \"Model\": model_name,\n",
    "        \"Accuracy\": accuracy,\n",
    "        \"Precision\": precision,\n",
    "        \"Recall\": recall,\n",
    "        \"F1 Score\": f1,\n",
    "        \"Time (seconds)\": elapsed_time\n",
    "    })\n",
    "    \n",
    "    # Tampilkan hasil classification report untuk setiap model\n",
    "    print(f\"\\nClassification Report untuk {model_name}:\")\n",
    "    print(classification_report(y_test, y_pred))\n",
    "\n",
    "dff = pd.DataFrame(results)\n",
    "dff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6566a66f-916c-4257-9497-ca916bd14982",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model</th>\n",
       "      <th>Accuracy</th>\n",
       "      <th>Precision</th>\n",
       "      <th>Recall</th>\n",
       "      <th>F1 Score</th>\n",
       "      <th>Time (seconds)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Random Forest</td>\n",
       "      <td>0.77280</td>\n",
       "      <td>0.772993</td>\n",
       "      <td>0.77280</td>\n",
       "      <td>0.770303</td>\n",
       "      <td>139.039397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Logistic Regression</td>\n",
       "      <td>0.65310</td>\n",
       "      <td>0.654736</td>\n",
       "      <td>0.65310</td>\n",
       "      <td>0.648167</td>\n",
       "      <td>144.464875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Support Vector Machine</td>\n",
       "      <td>0.10615</td>\n",
       "      <td>0.082078</td>\n",
       "      <td>0.10615</td>\n",
       "      <td>0.043710</td>\n",
       "      <td>4011.303041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>K-Nearest Neighbors</td>\n",
       "      <td>0.56100</td>\n",
       "      <td>0.568342</td>\n",
       "      <td>0.56100</td>\n",
       "      <td>0.559343</td>\n",
       "      <td>48.983661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MLP Classifier</td>\n",
       "      <td>0.83000</td>\n",
       "      <td>0.835395</td>\n",
       "      <td>0.83000</td>\n",
       "      <td>0.830154</td>\n",
       "      <td>615.483605</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Model  Accuracy  Precision   Recall  F1 Score  \\\n",
       "0           Random Forest   0.77280   0.772993  0.77280  0.770303   \n",
       "1     Logistic Regression   0.65310   0.654736  0.65310  0.648167   \n",
       "2  Support Vector Machine   0.10615   0.082078  0.10615  0.043710   \n",
       "3     K-Nearest Neighbors   0.56100   0.568342  0.56100  0.559343   \n",
       "4          MLP Classifier   0.83000   0.835395  0.83000  0.830154   \n",
       "\n",
       "   Time (seconds)  \n",
       "0      139.039397  \n",
       "1      144.464875  \n",
       "2     4011.303041  \n",
       "3       48.983661  \n",
       "4      615.483605  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hasil = pd.concat([results_df, dff], ignore_index=True)\n",
    "hasil"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c2145be4-c8a4-4ff9-a721-354696fa090d",
   "metadata": {},
   "outputs": [],
   "source": [
    "hasil.to_csv('../../result/Kmers+abundance_10fam.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afcd99b4-a614-4b57-977f-4211ca99983c",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Reads"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce8b4835-ed2e-42b7-b1e3-20db3bb78507",
   "metadata": {},
   "source": [
    "## Feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "05e558b8-1767-4d8f-87f8-babe392f81fe",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: './reads/anonymous_reads.fq.gz'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[12], line 6\u001b[0m\n\u001b[1;32m      3\u001b[0m length_threshold \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m100\u001b[39m  \u001b[38;5;66;03m# Example sequence length threshold\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m# Generate spaced k-mer features\u001b[39;00m\n\u001b[0;32m----> 6\u001b[0m featureR_df \u001b[38;5;241m=\u001b[39m \u001b[43mgenerate_kmer_features\u001b[49m\u001b[43m(\u001b[49m\u001b[43mread_file\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlength_threshold\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkmer_len\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      7\u001b[0m \u001b[38;5;28mprint\u001b[39m(featureR_df)\n",
      "Cell \u001b[0;32mIn[9], line 76\u001b[0m, in \u001b[0;36mgenerate_kmer_features\u001b[0;34m(folder_path, length_threshold, kmer_len, split, split_threshold)\u001b[0m\n\u001b[1;32m     74\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mgenerate_kmer_features\u001b[39m(folder_path, length_threshold, kmer_len, split\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m, split_threshold\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m):\n\u001b[1;32m     75\u001b[0m     \u001b[38;5;66;03m# List all files in the folder with .fna.gz extension\u001b[39;00m\n\u001b[0;32m---> 76\u001b[0m     files \u001b[38;5;241m=\u001b[39m [os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(folder_path, f) \u001b[38;5;28;01mfor\u001b[39;00m f \u001b[38;5;129;01min\u001b[39;00m os\u001b[38;5;241m.\u001b[39mlistdir(folder_path) \u001b[38;5;28;01mif\u001b[39;00m f\u001b[38;5;241m.\u001b[39mendswith(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.fna.gz\u001b[39m\u001b[38;5;124m'\u001b[39m)]\n\u001b[1;32m     78\u001b[0m     kmer_dict, kmer_names, nr_features \u001b[38;5;241m=\u001b[39m generate_feature_mapping(kmer_len)\n\u001b[1;32m     79\u001b[0m     composition \u001b[38;5;241m=\u001b[39m OrderedDict()\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: './reads/anonymous_reads.fq.gz'"
     ]
    }
   ],
   "source": [
    "read_file = './reads/anonymous_reads.fq.gz'  # Updated file path to your uploaded file\n",
    "kmer_len = 4  # Original k-mer length\n",
    "length_threshold = 100  # Example sequence length threshold\n",
    "\n",
    "# Generate spaced k-mer features\n",
    "featureR_df = generate_kmer_features(read_file, length_threshold, kmer_len)\n",
    "print(featureR_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5e7e656b-6d3f-4118-8f77-f74c94c53bad",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: './reads/reads_mapping.tsv.gz'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[13], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m labelR_path \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m./reads/reads_mapping.tsv.gz\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m----> 2\u001b[0m labelR_df \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlabelR_path\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      3\u001b[0m labelR_df \u001b[38;5;241m=\u001b[39m labelR_df[[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgenome_id\u001b[39m\u001b[38;5;124m\"\u001b[39m]]\n\u001b[1;32m      4\u001b[0m labelR_df\n",
      "File \u001b[0;32m/opt/conda/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1026\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m   1013\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m   1014\u001b[0m     dialect,\n\u001b[1;32m   1015\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1022\u001b[0m     dtype_backend\u001b[38;5;241m=\u001b[39mdtype_backend,\n\u001b[1;32m   1023\u001b[0m )\n\u001b[1;32m   1024\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m-> 1026\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/conda/lib/python3.11/site-packages/pandas/io/parsers/readers.py:620\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    617\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    619\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 620\u001b[0m parser \u001b[38;5;241m=\u001b[39m \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    622\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    623\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m/opt/conda/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1620\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1617\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1619\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1620\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/conda/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1880\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1878\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m mode:\n\u001b[1;32m   1879\u001b[0m         mode \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m-> 1880\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;241m=\u001b[39m \u001b[43mget_handle\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1881\u001b[0m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1882\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1883\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1884\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcompression\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1885\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmemory_map\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmemory_map\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1886\u001b[0m \u001b[43m    \u001b[49m\u001b[43mis_text\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mis_text\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1887\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mencoding_errors\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstrict\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1888\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mstorage_options\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1889\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1890\u001b[0m \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m   1891\u001b[0m f \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles\u001b[38;5;241m.\u001b[39mhandle\n",
      "File \u001b[0;32m/opt/conda/lib/python3.11/site-packages/pandas/io/common.py:765\u001b[0m, in \u001b[0;36mget_handle\u001b[0;34m(path_or_buf, mode, encoding, compression, memory_map, is_text, errors, storage_options)\u001b[0m\n\u001b[1;32m    761\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m compression \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgzip\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m    762\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(handle, \u001b[38;5;28mstr\u001b[39m):\n\u001b[1;32m    763\u001b[0m         \u001b[38;5;66;03m# error: Incompatible types in assignment (expression has type\u001b[39;00m\n\u001b[1;32m    764\u001b[0m         \u001b[38;5;66;03m# \"GzipFile\", variable has type \"Union[str, BaseBuffer]\")\u001b[39;00m\n\u001b[0;32m--> 765\u001b[0m         handle \u001b[38;5;241m=\u001b[39m \u001b[43mgzip\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mGzipFile\u001b[49m\u001b[43m(\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;66;43;03m# type: ignore[assignment]\u001b[39;49;00m\n\u001b[1;32m    766\u001b[0m \u001b[43m            \u001b[49m\u001b[43mfilename\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mhandle\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    767\u001b[0m \u001b[43m            \u001b[49m\u001b[43mmode\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mioargs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    768\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mcompression_args\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    769\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    770\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    771\u001b[0m         handle \u001b[38;5;241m=\u001b[39m gzip\u001b[38;5;241m.\u001b[39mGzipFile(\n\u001b[1;32m    772\u001b[0m             \u001b[38;5;66;03m# No overload variant of \"GzipFile\" matches argument types\u001b[39;00m\n\u001b[1;32m    773\u001b[0m             \u001b[38;5;66;03m# \"Union[str, BaseBuffer]\", \"str\", \"Dict[str, Any]\"\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    776\u001b[0m             \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mcompression_args,\n\u001b[1;32m    777\u001b[0m         )\n",
      "File \u001b[0;32m/opt/conda/lib/python3.11/gzip.py:174\u001b[0m, in \u001b[0;36mGzipFile.__init__\u001b[0;34m(self, filename, mode, compresslevel, fileobj, mtime)\u001b[0m\n\u001b[1;32m    172\u001b[0m     mode \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mb\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m    173\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m fileobj \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 174\u001b[0m     fileobj \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmyfileobj \u001b[38;5;241m=\u001b[39m builtins\u001b[38;5;241m.\u001b[39mopen(filename, mode \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrb\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m    175\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m filename \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    176\u001b[0m     filename \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mgetattr\u001b[39m(fileobj, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mname\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: './reads/reads_mapping.tsv.gz'"
     ]
    }
   ],
   "source": [
    "labelR_path = \"./reads/reads_mapping.tsv.gz\"\n",
    "labelR_df = pd.read_csv(labelR_path, sep='\\t')\n",
    "labelR_df = labelR_df[[\"genome_id\"]]\n",
    "labelR_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a2b55c96-5613-4532-837b-a5a82addf271",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'featureR_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[14], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m features \u001b[38;5;241m=\u001b[39m \u001b[43mfeatureR_df\u001b[49m\n\u001b[1;32m      2\u001b[0m labels \u001b[38;5;241m=\u001b[39m labelR_df\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m# Pisahkan data menjadi training dan testing set (80% training, 20% testing)\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'featureR_df' is not defined"
     ]
    }
   ],
   "source": [
    "features = featureR_df\n",
    "labels = labelR_df\n",
    "\n",
    "# Pisahkan data menjadi training dan testing set (80% training, 20% testing)\n",
    "X_train, X_test, y_train, y_test = train_test_split(features, labels, test_size=0.2, random_state=42)\n",
    "\n",
    "print(\"Jumlah data latih per kelas:\")\n",
    "print(np.unique(y_train, return_counts=True))\n",
    "\n",
    "print(\"\\nJumlah data uji per kelas:\")\n",
    "print(np.unique(y_test, return_counts=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6790daae-39bf-49e3-8723-724c4245e1cd",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'RandomForestClassifier' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[15], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;66;03m# Inisialisasi model Random Forest\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m rf_model \u001b[38;5;241m=\u001b[39m \u001b[43mRandomForestClassifier\u001b[49m(n_estimators\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m100\u001b[39m, random_state\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m42\u001b[39m)\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m# Latih model\u001b[39;00m\n\u001b[1;32m      5\u001b[0m rf_model\u001b[38;5;241m.\u001b[39mfit(X_train, y_train)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'RandomForestClassifier' is not defined"
     ]
    }
   ],
   "source": [
    "# Inisialisasi model Random Forest\n",
    "rf_model = RandomForestClassifier(n_estimators=100, random_state=42)\n",
    "\n",
    "# Latih model\n",
    "rf_model.fit(X_train, y_train)\n",
    "\n",
    "# Prediksi menggunakan data testing\n",
    "y_pred = rf_model.predict(X_test)\n",
    "\n",
    "# Tampilkan hasil akurasi dan classification report\n",
    "print(\"\\nAkurasi Model:\")\n",
    "print(accuracy_score(y_test, y_pred))\n",
    "\n",
    "print(\"\\nClassification Report:\")\n",
    "print(classification_report(y_test, y_pred))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8238cf9b-f486-4053-9b32-804b56bfd206",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
